{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting tidytcells\n",
      "  Using cached tidytcells-2.1.3-py3-none-any.whl.metadata (3.7 kB)\n",
      "Using cached tidytcells-2.1.3-py3-none-any.whl (90 kB)\n",
      "Installing collected packages: tidytcells\n",
      "Successfully installed tidytcells-2.1.3\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "!\"{sys.executable}\" -m pip install tidytcells\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set precision of mhc and V/J values (gene or allele)\n",
    "precision = 'allele'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# this function is not thread safe\n",
    "def create_folders_if_not_exists(folders):\n",
    "  for path in folders:\n",
    "    if not os.path.exists(path):\n",
    "      os.makedirs(path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "pipeline_data = './data_10x'\n",
    "pipeline_data_plain = f'{pipeline_data}/plain_datasets'\n",
    "pipeline_data_cleaned = f'{pipeline_data}/cleaned_datasets'\n",
    "pipeline_data_concatenated = f'{pipeline_data}/concatenated_datasets'\n",
    "pipeline_data_splitted = f'{pipeline_data}/splitted_datasets'\n",
    "pipeline_data_temp_bucket = f'{pipeline_data}/temp'\n",
    "\n",
    "pipeline_folders = [pipeline_data, pipeline_data_plain, pipeline_data_cleaned, pipeline_data_concatenated, pipeline_data_splitted, pipeline_data_temp_bucket]\n",
    "\n",
    "create_folders_if_not_exists(pipeline_folders)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Preparation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### IEDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# prepare directories\n",
    "IEDB_data_plain = f'{pipeline_data_plain}/IEDB'\n",
    "IEDB_data_cleaned = f'{pipeline_data_cleaned}/IEDB'\n",
    "IEDB_data_fitted = f'{pipeline_data_temp_bucket}/IEDB'\n",
    "\n",
    "IEDB_folders = [IEDB_data_plain, IEDB_data_cleaned, IEDB_data_fitted]\n",
    "create_folders_if_not_exists(IEDB_folders)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# prepare parameters for notebook IEDB fit data\n",
    "path_prefix_plain = IEDB_data_plain\n",
    "path_prefix_fitted = IEDB_data_fitted\n",
    "mhc_I_input_beta = f\"{path_prefix_plain}/MHCI_IEDB_beta_export.csv\"\n",
    "mhc_I_output_beta = f\"{path_prefix_fitted}/IEDB_beta_fitted.csv\"\n",
    "mhc_I_input_paired = f\"{path_prefix_plain}/MHCI_IEDB_paired_export.csv\"\n",
    "mhc_I_output_paired = f\"{path_prefix_fitted}/IEDB_paired_fitted.csv\"\n",
    "\n",
    "# fit IEDB data\n",
    "#%run ./data_scripts/IEDB/IEDB_fitted_dataset.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# prepare parameters for notebook IEDB clean data\n",
    "path_prefix_fitted = IEDB_data_fitted\n",
    "path_prefix_cleaned =  IEDB_data_cleaned\n",
    "fitted_file_beta = \"IEDB_beta_fitted.csv\"\n",
    "fitted_file_paired = \"IEDB_paired_fitted.csv\"\n",
    "cleaned_file_beta = \"IEDB_cleaned_data_beta.csv\"\n",
    "cleaned_file_paired = \"IEDB_cleaned_data_paired.csv\"\n",
    "\n",
    "# clean IEDB data\n",
    "#%run ./data_scripts/IEDB/IEDB_clean_dataset.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "IEDB_cleaned_beta_output = f'{IEDB_data_cleaned}/{cleaned_file_beta}'\n",
    "IEDB_cleaned_paired_output = f'{IEDB_data_cleaned}/{cleaned_file_paired}'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### McPAS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# prepare directories\n",
    "McPas_data_plain = f'{pipeline_data_plain}/McPas'\n",
    "McPas_data_cleaned = f'{pipeline_data_cleaned}/McPas'\n",
    "McPas_data_fitted = f'{pipeline_data_temp_bucket}/McPas'\n",
    "\n",
    "McPas_folders = [McPas_data_plain, McPas_data_cleaned, McPas_data_fitted]\n",
    "create_folders_if_not_exists(McPas_folders)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# prepare parameters for notebook McPAS fit data\n",
    "input_file = f'{McPas_data_plain}/McPAS-TCR.csv'\n",
    "path_prefix_fitted = McPas_data_fitted\n",
    "fitted_file = 'McPAS_fitted.tsv'\n",
    "\n",
    "# fit McPAS data\n",
    "%run ./data_scripts/McPas-TCR/fit_data_mcpastcr_both.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MHC Class I has 10078 entries\n",
      "whole dataframe has 13701 entries\n",
      "filtered to only use MHC Class I. Length of dataset: 10078\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_53158/2652240660.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "  mcpastcr_cleaned_both_df = mcpastcr_cleaned_both_df[~mask]\n"
     ]
    }
   ],
   "source": [
    "# prepare parameters for notebook McPAS clean data\n",
    "fitted_input_file = f'{McPas_data_fitted}/{fitted_file}'\n",
    "path_prefix_cleaned = McPas_data_cleaned\n",
    "cleaned_file_paired = 'McPAS_cleaned_data_paired.tsv'\n",
    "cleaned_file_beta = 'McPAS_cleaned_data_beta.tsv'\n",
    "\n",
    "# clean McPAS data\n",
    "%run ./data_scripts/McPas-TCR/clean_data_mcpastcr_both.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "McPAS_cleaned_beta_output = f'{McPas_data_cleaned}/{cleaned_file_beta}'\n",
    "McPAS_cleaned_paired_output = f'{McPas_data_cleaned}/{cleaned_file_paired}'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### VDJdb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# prepare directories\n",
    "VDJdb_data_plain = f'{pipeline_data_plain}/VDJdb'\n",
    "VDJdb_data_cleaned = f'{pipeline_data_cleaned}/VDJdb'\n",
    "VDJdb_data_fitted = f'{pipeline_data_temp_bucket}/VDJdb'\n",
    "\n",
    "VDJdb_folders = [VDJdb_data_plain, VDJdb_data_cleaned, VDJdb_data_fitted]\n",
    "create_folders_if_not_exists(VDJdb_folders)\n",
    "\n",
    "fitted_beta_file = 'VDJdb_beta_fitted.tsv'\n",
    "fitted_paired_file = 'VDJdb_paired_fitted.tsv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# prepare parameters for notebook VDJdb fit data paired\n",
    "input_file = f'{VDJdb_data_plain}/VDJdb_paired_only.tsv'\n",
    "path_prefix_fitted = VDJdb_data_fitted\n",
    "fitted_file = fitted_paired_file\n",
    "\n",
    "# fit paired VDJdb data\n",
    "%run ./data_scripts/VDJdb/fit_data_vdjdb_paired.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# prepare parameters for notebook VDJdb fit data beta\n",
    "input_file = f'{VDJdb_data_plain}/VDJdb_beta_only.tsv'\n",
    "path_prefix_fitted = VDJdb_data_fitted\n",
    "fitted_file = fitted_beta_file\n",
    "\n",
    "# fit beta VDJdb data\n",
    "%run ./data_scripts/VDJdb/fit_data_vdjdb_beta.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MHC Class I has 27414 entries\n",
      "whole dataframe has 28119 entries\n",
      "filtered to only use MHC Class I. Length of dataset: 27414\n"
     ]
    }
   ],
   "source": [
    "# prepare parameters for notebook VDJdb clean data paired\n",
    "input_file = f'{VDJdb_data_fitted}/{fitted_paired_file}'\n",
    "cleaned_file_paired = 'VDJdb_cleaned_data_paired.tsv'\n",
    "output_file = f'{VDJdb_data_cleaned}/{cleaned_file_paired}'\n",
    "\n",
    "# clean paired VDJdb data\n",
    "%run ./data_scripts/VDJdb/clean_data_vdjdb_paired.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MHC Class I has 46507 entries\n",
      "whole dataframe has 49042 entries\n",
      "filtered to only use MHC Class I. Length of dataset: 46507\n"
     ]
    }
   ],
   "source": [
    "# prepare parameters for notebook VDJdb clean data beta\n",
    "input_file = f'{VDJdb_data_fitted}/{fitted_beta_file}'\n",
    "cleaned_file_beta = 'VDJdb_cleaned_data_beta.tsv'\n",
    "output_file = f'{VDJdb_data_cleaned}/{cleaned_file_beta}'\n",
    "\n",
    "# clean beta VDJdb data\n",
    "%run ./data_scripts/VDJdb/clean_data_vdjdb_beta.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "VDJdb_cleaned_beta_output = f'{VDJdb_data_cleaned}/{cleaned_file_beta}'\n",
    "VDJdb_cleaned_paired_output = f'{VDJdb_data_cleaned}/{cleaned_file_paired}'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 10X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "' >>>>>>> alt\\nbase_path = \\'../data_10x/\\' \\n\\n# Einlesen der CSV-Dateien für die Donors\\ndonor1_consensus = pd.read_csv(base_path + \\'vdj_v1_hs_aggregated_donor1_consensus_annotations.csv\\')\\ndonor2_consensus = pd.read_csv(base_path + \\'vdj_v1_hs_aggregated_donor2_consensus_annotations.csv\\')\\ndonor3_consensus = pd.read_csv(base_path + \\'vdj_v1_hs_aggregated_donor3_consensus_annotations.csv\\')\\ndonor4_consensus = pd.read_csv(base_path + \\'vdj_v1_hs_aggregated_donor4_consensus_annotations.csv\\')\\n\\n# Daten prüfen\\nprint(\"Donor 1 Consensus:\", donor1_consensus.shape)\\nprint(\"Donor 2 Consensus:\", donor2_consensus.shape)\\nprint(\"Donor 3 Consensus:\", donor3_consensus.shape)\\nprint(\"Donor 4 Consensus:\", donor4_consensus.shape)\\n\\n\\n# Zusammenführen aller Donors in ein gemeinsames DataFrame\\nall_donors_consensus = pd.concat([donor1_consensus, donor2_consensus, donor3_consensus, donor4_consensus])\\n\\n# Optional: Reset index, falls nötig (nicht zwingend erforderlich, aber hilft bei der Verwaltung der Zeilenindices)\\nall_donors_consensus.reset_index(drop=True, inplace=True)\\n\\n# Gesamtstruktur anzeigen, um sicherzustellen, dass alles korrekt zusammengeführt wurde\\nprint(\"Kombinierte Donors:\", all_donors_consensus.shape)\\n\\n# Beispielhafter Blick auf die kombinierten Daten\\nprint(all_donors_consensus.head())'"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "''' >>>>>>> alt\n",
    "base_path = '../data_10x/' \n",
    "\n",
    "# Einlesen der CSV-Dateien für die Donors\n",
    "donor1_consensus = pd.read_csv(base_path + 'vdj_v1_hs_aggregated_donor1_consensus_annotations.csv')\n",
    "donor2_consensus = pd.read_csv(base_path + 'vdj_v1_hs_aggregated_donor2_consensus_annotations.csv')\n",
    "donor3_consensus = pd.read_csv(base_path + 'vdj_v1_hs_aggregated_donor3_consensus_annotations.csv')\n",
    "donor4_consensus = pd.read_csv(base_path + 'vdj_v1_hs_aggregated_donor4_consensus_annotations.csv')\n",
    "\n",
    "# Daten prüfen\n",
    "print(\"Donor 1 Consensus:\", donor1_consensus.shape)\n",
    "print(\"Donor 2 Consensus:\", donor2_consensus.shape)\n",
    "print(\"Donor 3 Consensus:\", donor3_consensus.shape)\n",
    "print(\"Donor 4 Consensus:\", donor4_consensus.shape)\n",
    "\n",
    "\n",
    "# Zusammenführen aller Donors in ein gemeinsames DataFrame\n",
    "all_donors_consensus = pd.concat([donor1_consensus, donor2_consensus, donor3_consensus, donor4_consensus])\n",
    "\n",
    "# Optional: Reset index, falls nötig (nicht zwingend erforderlich, aber hilft bei der Verwaltung der Zeilenindices)\n",
    "all_donors_consensus.reset_index(drop=True, inplace=True)\n",
    "\n",
    "# Gesamtstruktur anzeigen, um sicherzustellen, dass alles korrekt zusammengeführt wurde\n",
    "print(\"Kombinierte Donors:\", all_donors_consensus.shape)\n",
    "\n",
    "# Beispielhafter Blick auf die kombinierten Daten\n",
    "print(all_donors_consensus.head())'''\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"# Speichern der kombinierten Donor-Daten in einer CSV-Datei\\noutput_path = '../data_10x/combined_donors_consensus_annotations.csv'\\nall_donors_consensus.to_csv(output_path, index=False)\""
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''# Speichern der kombinierten Donor-Daten in einer CSV-Datei\n",
    "output_path = '../data_10x/combined_donors_consensus_annotations.csv'\n",
    "all_donors_consensus.to_csv(output_path, index=False)'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Donor 1 Consensus: (46526, 118)\n",
      "Donor 2 Consensus: (77854, 118)\n",
      "Donor 3 Consensus: (37824, 118)\n",
      "Donor 4 Consensus: (27308, 118)\n",
      "Kombinierte Donors: (189512, 118)\n",
      "               barcode   donor  \\\n",
      "0   AAACCTGAGACAAAGG-4  donor1   \n",
      "1  AAACCTGAGACTGTAA-34  donor1   \n",
      "2   AAACCTGAGAGCCCAA-5  donor1   \n",
      "3  AAACCTGAGAGCTGCA-24  donor1   \n",
      "4   AAACCTGAGAGGGATA-8  donor1   \n",
      "\n",
      "                                  cell_clono_cdr3_aa  \\\n",
      "0  TRA:CAASVSIWTGTASKLTF;TRA:CAAWDMEYGNKLVF;TRB:C...   \n",
      "1                                    TRB:CASDTPVGQFF   \n",
      "2                 TRA:CASYTDKLIF;TRB:CASSGGSISTDTQYF   \n",
      "3                                 TRB:CASSGGQSSYEQYF   \n",
      "4          TRA:CAASGYGNTGRRALTF;TRB:CASSQDPAGGYNEQFF   \n",
      "\n",
      "                                  cell_clono_cdr3_nt     CD3  CD19  CD45RA  \\\n",
      "0  TRA:TGTGCAGCAAGCGTTAGTATTTGGACCGGCACTGCCAGTAAA...  2125.0   0.0   912.0   \n",
      "1              TRB:TGTGCCAGCGATACCCCGGTTGGGCAGTTCTTC  1023.0   0.0  2028.0   \n",
      "2  TRA:TGTGCTTCCTACACCGACAAGCTCATCTTT;TRB:TGCGCCA...  1598.0   3.0  3454.0   \n",
      "3     TRB:TGCGCCAGCAGTGGCGGACAGAGCTCCTACGAGCAGTACTTC   298.0   1.0   880.0   \n",
      "4  TRA:TGTGCAGCAAGCGGGTATGGAAACACGGGCAGGAGAGCACTT...  1036.0   0.0  2457.0   \n",
      "\n",
      "   CD4    CD8a  CD14  ...  B0702_RPHERNGFTVL_pp65_CMV_binder  \\\n",
      "0  1.0  2223.0   4.0  ...                              False   \n",
      "1  2.0  3485.0   1.0  ...                              False   \n",
      "2  4.0  3383.0   1.0  ...                              False   \n",
      "3  1.0  2389.0   1.0  ...                              False   \n",
      "4  2.0  3427.0   3.0  ...                              False   \n",
      "\n",
      "   B0801_RAKFKQLL_BZLF1_EBV_binder  B0801_ELRRKMMYM_IE-1_CMV_binder  \\\n",
      "0                            False                            False   \n",
      "1                            False                            False   \n",
      "2                            False                            False   \n",
      "3                            False                            False   \n",
      "4                            False                            False   \n",
      "\n",
      "   B0801_FLRGRAYGL_EBNA-3A_EBV_binder  A0101_SLEGGGLGY_NC_binder  \\\n",
      "0                               False                      False   \n",
      "1                               False                      False   \n",
      "2                               False                      False   \n",
      "3                               False                      False   \n",
      "4                               False                      False   \n",
      "\n",
      "   A0101_STEGGGLAY_NC_binder  A0201_ALIAPVHAV_NC_binder  \\\n",
      "0                      False                      False   \n",
      "1                      False                      False   \n",
      "2                      False                      False   \n",
      "3                      False                      False   \n",
      "4                      False                      False   \n",
      "\n",
      "   A2402_AYSSAGASI_NC_binder  B0702_GPAESAAGL_NC_binder  \\\n",
      "0                      False                      False   \n",
      "1                      False                      False   \n",
      "2                      False                      False   \n",
      "3                      False                      False   \n",
      "4                      False                      False   \n",
      "\n",
      "   NR(B0801)_AAKGRGAAL_NC_binder  \n",
      "0                          False  \n",
      "1                          False  \n",
      "2                          False  \n",
      "3                          False  \n",
      "4                          False  \n",
      "\n",
      "[5 rows x 118 columns]\n",
      "Anzahl der entfernten Duplikate: 0\n",
      "Kombinierte Donors nach Duplikatentfernung: (189512, 118)\n"
     ]
    }
   ],
   "source": [
    "### NEU\n",
    "\n",
    "base_path = '../data_10x/' \n",
    "\n",
    "# Einlesen der CSV-Dateien für die Donors\n",
    "donor1_consensus = pd.read_csv(base_path + 'vdj_v1_hs_aggregated_donor1_binarized_matrix.csv')\n",
    "donor2_consensus = pd.read_csv(base_path + 'vdj_v1_hs_aggregated_donor2_binarized_matrix.csv')\n",
    "donor3_consensus = pd.read_csv(base_path + 'vdj_v1_hs_aggregated_donor3_binarized_matrix.csv')\n",
    "donor4_consensus = pd.read_csv(base_path + 'vdj_v1_hs_aggregated_donor4_binarized_matrix.csv')\n",
    "\n",
    "# Daten prüfen\n",
    "print(\"Donor 1 Consensus:\", donor1_consensus.shape)\n",
    "print(\"Donor 2 Consensus:\", donor2_consensus.shape)\n",
    "print(\"Donor 3 Consensus:\", donor3_consensus.shape)\n",
    "print(\"Donor 4 Consensus:\", donor4_consensus.shape)\n",
    "\n",
    "\n",
    "# Zusammenführen aller Donors in ein gemeinsames DataFrame\n",
    "all_donors_consensus = pd.concat([donor1_consensus, donor2_consensus, donor3_consensus, donor4_consensus])\n",
    "\n",
    "# Optional: Reset index, falls nötig (nicht zwingend erforderlich, aber hilft bei der Verwaltung der Zeilenindices)\n",
    "all_donors_consensus.reset_index(drop=True, inplace=True)\n",
    "\n",
    "# Gesamtstruktur anzeigen, um sicherzustellen, dass alles korrekt zusammengeführt wurde\n",
    "print(\"Kombinierte Donors:\", all_donors_consensus.shape)\n",
    "\n",
    "# Beispielhafter Blick auf die kombinierten Daten\n",
    "print(all_donors_consensus.head())\n",
    "\n",
    "# Entfernen von Duplikaten\n",
    "initial_count = all_donors_consensus.shape[0]\n",
    "all_donors_consensus.drop_duplicates(inplace=True)\n",
    "final_count = all_donors_consensus.shape[0]\n",
    "duplicates_removed = initial_count - final_count\n",
    "\n",
    "# Anzahl der Duplikate anzeigen\n",
    "print(\"Anzahl der entfernten Duplikate:\", duplicates_removed)\n",
    "print(\"Kombinierte Donors nach Duplikatentfernung:\", all_donors_consensus.shape)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Speichern der kombinierten Donor-Daten in einer CSV-Datei\n",
    "output_path = '../data_10x/combined_donors_consensus_annotations.csv'\n",
    "all_donors_consensus.to_csv(output_path, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_53158/3438921542.py:7: DtypeWarning: Columns (16,17) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  donor2_meta = pd.read_csv(base_path + 'vdj_v1_hs_aggregated_donor2_all_contig_annotations.csv')\n",
      "/tmp/ipykernel_53158/3438921542.py:8: DtypeWarning: Columns (16,17) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  donor3_meta = pd.read_csv(base_path + 'vdj_v1_hs_aggregated_donor3_all_contig_annotations.csv')\n",
      "/tmp/ipykernel_53158/3438921542.py:9: DtypeWarning: Columns (16,17) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  donor4_meta = pd.read_csv(base_path + 'vdj_v1_hs_aggregated_donor4_all_contig_annotations.csv')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Donor 1: (187144, 18)\n",
      "Donor 2: (299037, 18)\n",
      "Donor 3: (429296, 18)\n",
      "Donor 4: (309855, 18)\n",
      "Kombinierte Donors: (1225332, 18)\n",
      "              barcode  is_cell                    contig_id  high_confidence  \\\n",
      "0  AAACCTGAGACAAAGG-4     True  AAACCTGAGACAAAGG-4_contig_1             True   \n",
      "1  AAACCTGAGACAAAGG-4     True  AAACCTGAGACAAAGG-4_contig_2             True   \n",
      "2  AAACCTGAGACAAAGG-4     True  AAACCTGAGACAAAGG-4_contig_3             True   \n",
      "3  AAACCTGAGACAAAGG-4     True  AAACCTGAGACAAAGG-4_contig_4            False   \n",
      "4  AAACCTGAGACAAAGG-4     True  AAACCTGAGACAAAGG-4_contig_5            False   \n",
      "\n",
      "   length chain     v_gene d_gene   j_gene c_gene  full_length productive  \\\n",
      "0     722   TRB   TRBV10-3  TRBD2  TRBJ2-1  TRBC2         True       True   \n",
      "1     605   TRA  TRAV29DV5    NaN   TRAJ44   TRAC         True       True   \n",
      "2     738   TRA    TRAV8-6    NaN   TRAJ47   TRAC         True       True   \n",
      "3     468   TRB        NaN    NaN  TRBJ2-3  TRBC2        False        NaN   \n",
      "4     488   TRB        NaN    NaN  TRBJ2-6  TRBC2        False        NaN   \n",
      "\n",
      "                cdr3                                            cdr3_nt  \\\n",
      "0  CAISDPGLAGGGGEQFF  TGTGCCATCAGTGACCCCGGACTAGCGGGAGGCGGGGGGGAGCAGT...   \n",
      "1  CAASVSIWTGTASKLTF  TGTGCAGCAAGCGTTAGTATTTGGACCGGCACTGCCAGTAAACTCA...   \n",
      "2     CAAWDMEYGNKLVF         TGTGCCGCCTGGGACATGGAATATGGAAACAAGCTGGTCTTT   \n",
      "3                NaN                                                NaN   \n",
      "4                NaN                                                NaN   \n",
      "\n",
      "   reads  umis raw_clonotype_id         raw_consensus_id  \n",
      "0  32237    18      clonotype19  clonotype19_consensus_1  \n",
      "1   6088     3      clonotype19  clonotype19_consensus_2  \n",
      "2   5358     3      clonotype19  clonotype19_consensus_3  \n",
      "3   2517     1      clonotype19                      NaN  \n",
      "4   2468     1      clonotype19                      NaN  \n",
      "Anzahl der entfernten Duplikate: 252\n",
      "Kombinierte Donors nach Duplikatentfernung: (1225080, 18)\n"
     ]
    }
   ],
   "source": [
    "#zweiter Datensatz v, d, j, c Infos\n",
    "\n",
    "base_path = '../data_10x/' \n",
    "\n",
    "# Einlesen der CSV-Dateien für die Donors\n",
    "donor1_meta = pd.read_csv(base_path + 'vdj_v1_hs_aggregated_donor1_all_contig_annotations.csv')\n",
    "donor2_meta = pd.read_csv(base_path + 'vdj_v1_hs_aggregated_donor2_all_contig_annotations.csv')\n",
    "donor3_meta = pd.read_csv(base_path + 'vdj_v1_hs_aggregated_donor3_all_contig_annotations.csv')\n",
    "donor4_meta = pd.read_csv(base_path + 'vdj_v1_hs_aggregated_donor4_all_contig_annotations.csv')\n",
    "\n",
    "# Daten prüfen\n",
    "print(\"Donor 1:\", donor1_meta.shape)\n",
    "print(\"Donor 2:\", donor2_meta.shape)\n",
    "print(\"Donor 3:\", donor3_meta.shape)\n",
    "print(\"Donor 4:\", donor4_meta.shape)\n",
    "\n",
    "\n",
    "# Zusammenführen aller Donors in ein gemeinsames DataFrame\n",
    "all_donors_meta = pd.concat([donor1_meta, donor2_meta, donor3_meta, donor4_meta])\n",
    "\n",
    "# Optional: Reset index, falls nötig (nicht zwingend erforderlich, aber hilft bei der Verwaltung der Zeilenindices)\n",
    "all_donors_meta.reset_index(drop=True, inplace=True)\n",
    "\n",
    "# Gesamtstruktur anzeigen, um sicherzustellen, dass alles korrekt zusammengeführt wurde\n",
    "print(\"Kombinierte Donors:\", all_donors_meta.shape)\n",
    "\n",
    "# Beispielhafter Blick auf die kombinierten Daten\n",
    "print(all_donors_meta.head())\n",
    "\n",
    "# Entfernen von Duplikaten\n",
    "initial_count = all_donors_meta.shape[0]\n",
    "all_donors_meta.drop_duplicates(inplace=True)\n",
    "final_count = all_donors_meta.shape[0]\n",
    "duplicates_removed = initial_count - final_count\n",
    "\n",
    "# Anzahl der Duplikate anzeigen\n",
    "print(\"Anzahl der entfernten Duplikate:\", duplicates_removed)\n",
    "print(\"Kombinierte Donors nach Duplikatentfernung:\", all_donors_meta.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Speichern der kombinierten Donor-Daten in einer CSV-Datei\n",
    "output_path = '../data_10x/meta.csv'\n",
    "all_donors_meta.to_csv(output_path, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Concatenation\n",
    "The concatenation includes further cleaning and advanced removal of duplicated rows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "length of beta_df: 231627\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"H-2Kb\" for species homosapiens: unrecognised gene name. (best attempted fix: \"HLA-H-2KB\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"HLA-A*24:01\" for species homosapiens: nonexistent allele for recognised gene.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"HLA-B*12\" for species homosapiens: nonexistent allele for recognised gene.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"HLA-A*08:01\" for species homosapiens: nonexistent allele for recognised gene.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"HLA class I\" for species homosapiens: unrecognised gene name. (best attempted fix: \"HLA-HLACLASSI\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBV9-02\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRBV9-2\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBV2-7*01\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBV19*04\" for species homosapiens: nonexistent allele for recognised gene.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBV26-2*01\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBV19*05\" for species homosapiens: nonexistent allele for recognised gene.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBV26-1*03\" for species homosapiens: nonexistent allele for recognised gene. (best attempted fix: \"TRBV26*03\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBV2-6*01\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBV8-4*05\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBV2-2*01\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBV1-4*01\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBV2-03\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRBV2-3\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBV2-05\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRBV2-5\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBV13-2\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBV13-06\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRBV13-6\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBV1-05\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRBV1-5\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBV21-03\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRBV21-3\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBV6-2*02\" for species homosapiens: nonexistent allele for recognised gene.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TCRB17-1\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRB17-1\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TCRBV21-3\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRBV21-3\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TCRBV13-6\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRBV13-6\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TCRBV13-3\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRBV13-3\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TCRVB06\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRVB6\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TCRVB6\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRVB6\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TCRVB12\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRVB12\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TCRBV12-3/4\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRBV12-3/4\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TCRBB27\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRBB27\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"05-07*01\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TR5-7*01\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ5-1\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ10-1\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ20-1\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ19-1\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ17-1\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ38-2*01\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"Donor 13\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRDON/OR13\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"Negative\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRNEGATIVE\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ37*02\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ8*01\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ52*01\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ9*01\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ33*01\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ36*01\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ53*01\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ54*01\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ3-1*01\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ42*01\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ39*01\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ45*01\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ31*01\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ43*01\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ1-7\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ5-6\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ3-2\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize Donor13: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CALQDXNTGEXFF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CSARTGDRTEAFX: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSILGWSEAFX: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSLRTRTDTQYX: not a valid amino acid sequence.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The following script removes a lot of rows. They are kept and some of them get added again later\n",
      "distinct entries (all columns, keep=first). 36836 entries removed.\n",
      "removed all duplicates (CDR3, Epitope) from distinct values (most_important_columns, keep=False). 47581 entries removed.\n",
      "beta removed entries df length: 47581\n",
      "\n",
      "\n",
      "Number of groups formed: 18337\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_53158/3398904298.py:24: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  duplicates_to_add = pd.concat([duplicates_to_add, group[group['is_duplicated'] == False]])\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32617 can be re-added to the no-duplicated dataframe\n",
      "from the plain dataset which has 231622 entries, 51800 entries have been removed.\n",
      "for beta dataset :\n",
      "size difference is: 51800\n",
      "  179822 information score cleaned: 5.202255563835348\n",
      "  231622 information score dropout: 5.0485964200291855\n",
      "final_beta_df length = 179822\n",
      "length of paired_df: 54338\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"H-2Kb\" for species homosapiens: unrecognised gene name. (best attempted fix: \"HLA-H-2KB\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"HLA-A*24:01\" for species homosapiens: nonexistent allele for recognised gene.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"HLA-A*08:01\" for species homosapiens: nonexistent allele for recognised gene.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"HLA-B*12\" for species homosapiens: nonexistent allele for recognised gene.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"HLA class I\" for species homosapiens: unrecognised gene name. (best attempted fix: \"HLA-HLACLASSI\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRAV1-4\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRAV2-2\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRAV251\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRAV21-2\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRDAV1*01\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRA21-01\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRA21-1\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRAV19-*01\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TCRAV1-4\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRAV1-4\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TCRAV13/1*02 (F)\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRAV13/DV1*02\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TCRAV2-2\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRAV2-2\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TCRAV9*02\" for species homosapiens: nonexistent allele for recognised gene. (best attempted fix: \"TRAV9-1*02\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRAV12-4\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TCRAV24*0\" for species homosapiens: nonexistent allele for recognised gene. (best attempted fix: \"TRAV24*00\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TCRAV12-4\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRAV12-4\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBV9-02\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRBV9-2\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TCRBV13-3\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRBV13-3\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"CATSESSGQTYEQYF\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRCAT-E--GQTYEQYF\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRAJ53*02\" for species homosapiens: nonexistent allele for recognised gene.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"Negative\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRNEGATIVE\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"Donor 13\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRDON/OR13\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ5-6\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ5-1\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ10-1\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ20-1\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ38-2*01\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ17-1\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ19-1\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize TRAJ37: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize XVXNREDKLVF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CVVNNNXDMRF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CVVNGMDSSYKLXF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize VQQDLGY##GGTSYGKLTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize VL*#GGGADGLTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CAVED#ISSGSARQLTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CIVRQLGTGAVVTIN*#F: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CTSVQQAL#GGSQGNLIF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CIVSVLG#SGGSNYKLTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize XEXXGGXXRXXX##: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CI##GGADGLTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CTPG#SQGNLIF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CIVRV#NSNSGYALNF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CAVSA#KAAGNKLTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CAVSAAX#AGNKLTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CAVSGPS#TDSWGKLQF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CAPQTRRLATTVS*#W: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CAGLGGT#GGSNYKLTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CAVND#NTDKLIF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize VL*VLPG#ALNF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CAIELQAR##: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize VQQAQ#YGGSQGNLIF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CAASSWS*#GLTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CAVEDRGEKHXSCL: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CAVERGL##GSQGNLIF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CAVS#SNFGNEKLTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CALPP#YNFNKFYF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CAXSGGGSYQLTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CAG#YGNKLVF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CIVRVA#SGNTPLVF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CIXXL*NDIGENMFLI: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CISP*SGSSNTGKLIF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize VQQAG#SNYKLTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CATLCPQ#NKLTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CAASGXGTYKYIF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CAXRGGSEKLVF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CAVN?PPFGNEKLTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize Donor13: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSILGWSEAFX: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CSARTGDRTEAFX: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSLRTRTDTQYX: not a valid amino acid sequence.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The following script removes a lot of rows. They are kept and some of them get added again later\n",
      "distinct entries (all columns, keep=first). 6090 entries removed.\n",
      "removed all duplicates from distinct values (cultivated columns, keep=False). 32381 entries removed.\n",
      "paired removed entries df length: 32381\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_53158/2624603129.py:18: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  duplicates_to_add = pd.concat([duplicates_to_add, group[group['is_duplicated'] == False]])\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32337 can be re-added to the no-duplicated dataframe\n",
      "from the plain dataset which has 54295 entries, 6134 entries have been removed.\n",
      "for paired dataset:\n",
      "size difference is: 6134\n",
      "  48161 information score cleaned: 7.147442951765952\n",
      "  54295 information score dropout: 7.243300488074408\n",
      "final_paired_df length: 48161\n"
     ]
    }
   ],
   "source": [
    "# prepare parameters for concatenation\n",
    "custom_dataset_path = f'{pipeline_data_concatenated}/{precision}/'\n",
    "\n",
    "# beta input files\n",
    "vdjdb_beta_read_path = VDJdb_cleaned_beta_output\n",
    "mcpastcr_beta_read_path = McPAS_cleaned_beta_output\n",
    "iedb_beta_read_path = IEDB_cleaned_beta_output\n",
    "# paired input files\n",
    "vdjdb_paired_read_path = VDJdb_cleaned_paired_output\n",
    "mcpastcr_paired_read_path = McPAS_cleaned_paired_output\n",
    "iedb_paired_read_path = IEDB_cleaned_paired_output\n",
    "# output files\n",
    "output_file_beta = 'beta_concatenated.tsv'\n",
    "output_file_paired = 'paired_concatenated.tsv'\n",
    "\n",
    "create_folders_if_not_exists([custom_dataset_path])\n",
    "\n",
    "%run ./data_scripts/concatDatasets.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# prepare parameters for concatenation\n",
    "custom_dataset_path = f'{pipeline_data_concatenated}/{precision}/'\n",
    "# output files\n",
    "output_file_beta = 'beta_concatenated.tsv'\n",
    "output_file_paired = 'paired_concatenated.tsv'\n",
    "\n",
    "concatenated_paired = f'{custom_dataset_path}/{output_file_paired}'\n",
    "concatenated_beta = f'{custom_dataset_path}/{output_file_beta}'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data split\n",
    "The split creates 3 datasets. Train, Validation and Test. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "distinct tcr's: 29339 from 48161\n",
      "unique tcr's: 13525 from 48161\n",
      "unique epitopes: 616 from 48161\n",
      "train data has 34636 entries\n",
      "test data has 13525 entries\n",
      "test data has 0 TPP1 tasks (unseen tcr & seen epitopes).\n",
      "test data has 11253 TPP2 tasks (unseen tcr & seen epitopes).\n",
      "test data has 2272 TPP3 tasks (unseen tcr & unseen epitope).\n",
      "the train/test ratio is 0.7191711135566121/0.2808288864433878\n",
      "924 entries need to be shifted from train to test so the train/test ratio can be 0.7/0.3\n",
      "924 entries from train will be moved to test (TPP1)\n",
      "df_train size before: 34636\n",
      "number of tpp1 before: 0\n",
      "number of tpp2 before: 11253\n",
      "df_train size after: 33712\n",
      "number of tpp1 after: 924\n",
      "number of tpp2 after: 11253\n",
      "5164 entries will be shifted from test to train so the tpp1/tpp2 ratio can be 0.5/0.5\n",
      "5165 entries need to be shifted from train to test so the tpp1/tpp2 ratio can be 0.5/0.5\n",
      "5165 entries from train will be moved to test (TPP1)\n",
      "df_train size before: 38876\n",
      "number of tpp1 before: 924\n",
      "number of tpp2 before: 6089\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_53158/4063483123.py:19: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  df_train = pd.concat([df_train, rows_to_move], ignore_index=True)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "df_train size after: 33711\n",
      "number of tpp1 after: 6089\n",
      "number of tpp2 after: 6089\n",
      "train data has 33711 entries\n",
      "test data has 14450 entries\n",
      "test data has 6089 TPP1 tasks (seen tcr & seen epitopes).\n",
      "test data has 6089 TPP2 tasks (unseen tcr & seen epitopes).\n",
      "test data has 2272 TPP3 tasks (unseen tcr & unseen epitope).\n",
      "the train/test ratio is 0.6999647017296152/0.3000352982703848\n",
      "test data has 7226 entries\n",
      "validation data has 7224 entries\n",
      "train data has 33711 entries\n",
      "test data has 3045 TPP1 tasks (seen tcr & seen epitopes).\n",
      "test data has 3891 TPP2 tasks (unseen tcr & seen epitopes).\n",
      "test data has 290 TPP3 tasks (unseen tcr & unseen epitope).\n",
      "the test ratio is 0.849961587176346/0.150038412823654\n",
      "the validation ratio is 0.8500031145532693/0.14999688544673076\n"
     ]
    }
   ],
   "source": [
    "# prepare parameters for split of paired dataset\n",
    "input_file = concatenated_paired\n",
    "paired_output_folder = f'{pipeline_data_splitted}/{precision}/paired'\n",
    "validation_file_name = 'validation.tsv'\n",
    "test_file_name = 'test.tsv'\n",
    "train_file_name = 'train.tsv'\n",
    "aimed_test_ratio = 0.3 # this means 30% of the concatenated dataset will be for test and validation (fifty/fifty)\n",
    "\n",
    "create_folders_if_not_exists([paired_output_folder])\n",
    "\n",
    "# do the split\n",
    "%run ./data_scripts/data_preparation/split_paired.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "distinct tcr's: 152160 from 179822\n",
      "unique tcr's: 139540 from 179822\n",
      "unique epitopes: 678 from 179822\n",
      "train data has 40282 entries\n",
      "test data has 139540 entries\n",
      "test data has 0 TPP1 tasks (unseen tcr & seen epitopes).\n",
      "test data has 137217 TPP2 tasks (unseen tcr & seen epitopes).\n",
      "test data has 2323 TPP3 tasks (unseen tcr & unseen epitope).\n",
      "the train/test ratio is 0.22401041029462465/0.7759895897053753\n",
      "85594 entries will be shifted from test to train so the train/test ratio can be 0.7/0.3\n",
      "25811 entries will be shifted from test to train so the tpp1/tpp2 ratio can be 0.5/0.5\n",
      "25812 entries need to be shifted from train to test so the tpp1/tpp2 ratio can be 0.5/0.5\n",
      "train data has 125875 entries\n",
      "test data has 53947 entries\n",
      "test data has 25812 TPP1 tasks (seen tcr & seen epitopes).\n",
      "test data has 25812 TPP2 tasks (unseen tcr & seen epitopes).\n",
      "test data has 2323 TPP3 tasks (unseen tcr & unseen epitope).\n",
      "the train/test ratio is 0.6999977755780716/0.30000222442192837\n",
      "test data has 26974 entries\n",
      "validation data has 26973 entries\n",
      "train data has 125875 entries\n",
      "test data has 12906 TPP1 tasks (seen tcr & seen epitopes).\n",
      "test data has 13764 TPP2 tasks (unseen tcr & seen epitopes).\n",
      "test data has 304 TPP3 tasks (unseen tcr & unseen epitope).\n",
      "the test ratio is 0.8499961072616253/0.1500038927383746\n",
      "the validation ratio is 0.8500016683164463/0.14999833168355373\n"
     ]
    }
   ],
   "source": [
    "# prepare parameters for split of beta dataset\n",
    "input_file = concatenated_beta\n",
    "beta_output_folder = f'{pipeline_data_splitted}/{precision}/beta'\n",
    "aimed_test_ratio = 0.3 # this means 30% of the concatenated dataset will be for test and validation (fifty/fifty)\n",
    "\n",
    "create_folders_if_not_exists([beta_output_folder])\n",
    "\n",
    "# do the split\n",
    "%run ./data_scripts/data_preparation/split_beta.ipynb"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Negative Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_53158/3854116297.py:9: DtypeWarning: Columns (16,17) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  all_donors_meta = pd.read_csv(all_donors_meta_path, sep=',')\n"
     ]
    }
   ],
   "source": [
    "#Daten einlesen\n",
    "\n",
    "combined_donors_path = '../data_10x/combined_donors_consensus_annotations.csv'\n",
    "all_donors_consensus = pd.read_csv(combined_donors_path, sep=',')\n",
    "\n",
    "#print(\"Consensus: \", all_donors_consensus.head())\n",
    "\n",
    "all_donors_meta_path = '../data_10x/meta.csv'\n",
    "all_donors_meta = pd.read_csv(all_donors_meta_path, sep=',')\n",
    "\n",
    "#print(\"Meta: \", all_donors_meta.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Beta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Batch Start:  0\n",
      "Batch Start:  1000\n",
      "Batch Start:  2000\n",
      "Batch Start:  3000\n",
      "Batch Start:  4000\n",
      "Batch Start:  5000\n",
      "Batch Start:  6000\n",
      "Batch Start:  7000\n",
      "Batch Start:  8000\n",
      "Batch Start:  9000\n",
      "Batch Start:  10000\n",
      "Batch Start:  11000\n",
      "Batch Start:  12000\n",
      "Batch Start:  13000\n",
      "Batch Start:  14000\n",
      "Batch Start:  15000\n",
      "Batch Start:  16000\n",
      "Batch Start:  17000\n",
      "Batch Start:  18000\n",
      "Batch Start:  19000\n",
      "Batch Start:  20000\n",
      "Batch Start:  21000\n",
      "Batch Start:  22000\n",
      "Batch Start:  23000\n",
      "Batch Start:  24000\n",
      "Batch Start:  25000\n",
      "Batch Start:  26000\n",
      "Batch Start:  27000\n",
      "Batch Start:  28000\n",
      "Batch Start:  29000\n",
      "Batch Start:  30000\n",
      "Batch Start:  31000\n",
      "Batch Start:  32000\n",
      "Batch Start:  33000\n",
      "Batch Start:  34000\n",
      "Batch Start:  35000\n",
      "Batch Start:  36000\n",
      "Batch Start:  37000\n",
      "Batch Start:  38000\n",
      "Batch Start:  39000\n",
      "Batch Start:  40000\n",
      "Batch Start:  41000\n",
      "Batch Start:  42000\n",
      "Batch Start:  43000\n",
      "Batch Start:  44000\n",
      "Batch Start:  45000\n",
      "Batch Start:  46000\n",
      "Batch Start:  47000\n",
      "Batch Start:  48000\n",
      "Batch Start:  49000\n",
      "Batch Start:  50000\n",
      "Batch Start:  51000\n",
      "Batch Start:  52000\n",
      "Batch Start:  53000\n",
      "Batch Start:  54000\n",
      "Batch Start:  55000\n",
      "Batch Start:  56000\n",
      "Batch Start:  57000\n",
      "Batch Start:  58000\n",
      "Batch Start:  59000\n",
      "Batch Start:  60000\n",
      "Batch Start:  61000\n",
      "Batch Start:  62000\n",
      "Batch Start:  63000\n",
      "Batch Start:  64000\n",
      "Batch Start:  65000\n",
      "Batch Start:  66000\n",
      "Batch Start:  67000\n",
      "Batch Start:  68000\n",
      "Batch Start:  69000\n",
      "Batch Start:  70000\n",
      "Batch Start:  71000\n",
      "Batch Start:  72000\n",
      "Batch Start:  73000\n",
      "Batch Start:  74000\n",
      "Batch Start:  75000\n",
      "Batch Start:  76000\n",
      "Batch Start:  77000\n",
      "Batch Start:  78000\n",
      "Batch Start:  79000\n",
      "Batch Start:  80000\n",
      "Batch Start:  81000\n",
      "Batch Start:  82000\n",
      "Batch Start:  83000\n",
      "Batch Start:  84000\n",
      "Batch Start:  85000\n",
      "Batch Start:  86000\n",
      "Batch Start:  87000\n",
      "Batch Start:  88000\n",
      "Batch Start:  89000\n",
      "Batch Start:  90000\n",
      "Batch Start:  91000\n",
      "Batch Start:  92000\n",
      "Batch Start:  93000\n",
      "Batch Start:  94000\n",
      "Batch Start:  95000\n",
      "Batch Start:  96000\n",
      "Batch Start:  97000\n",
      "Batch Start:  98000\n",
      "Batch Start:  99000\n",
      "Batch Start:  100000\n",
      "Batch Start:  101000\n",
      "Batch Start:  102000\n",
      "Batch Start:  103000\n",
      "Batch Start:  104000\n",
      "Batch Start:  105000\n",
      "Batch Start:  106000\n",
      "Batch Start:  107000\n",
      "Batch Start:  108000\n",
      "Batch Start:  109000\n",
      "Batch Start:  110000\n",
      "Batch Start:  111000\n",
      "Batch Start:  112000\n",
      "Batch Start:  113000\n",
      "Batch Start:  114000\n",
      "Batch Start:  115000\n",
      "Batch Start:  116000\n",
      "Batch Start:  117000\n",
      "Batch Start:  118000\n",
      "Batch Start:  119000\n",
      "Batch Start:  120000\n",
      "Batch Start:  121000\n",
      "Batch Start:  122000\n",
      "Batch Start:  123000\n",
      "Batch Start:  124000\n",
      "Batch Start:  125000\n",
      "Batch Start:  126000\n",
      "Batch Start:  127000\n",
      "Batch Start:  128000\n",
      "Batch Start:  129000\n",
      "Batch Start:  130000\n",
      "Batch Start:  131000\n",
      "Batch Start:  132000\n",
      "Batch Start:  133000\n",
      "Batch Start:  134000\n",
      "Batch Start:  135000\n",
      "Batch Start:  136000\n",
      "Batch Start:  137000\n",
      "Batch Start:  138000\n",
      "Batch Start:  139000\n",
      "Batch Start:  140000\n",
      "Batch Start:  141000\n",
      "Batch Start:  142000\n",
      "Batch Start:  143000\n",
      "Batch Start:  144000\n",
      "Batch Start:  145000\n",
      "Batch Start:  146000\n",
      "Batch Start:  147000\n",
      "Batch Start:  148000\n",
      "Batch Start:  149000\n",
      "Batch Start:  150000\n",
      "Batch Start:  151000\n",
      "Batch Start:  152000\n",
      "Batch Start:  153000\n",
      "Batch Start:  154000\n",
      "Batch Start:  155000\n",
      "Batch Start:  156000\n",
      "Batch Start:  157000\n",
      "Batch Start:  158000\n",
      "Batch Start:  159000\n",
      "Batch Start:  160000\n",
      "Batch Start:  161000\n",
      "Batch Start:  162000\n",
      "Batch Start:  163000\n",
      "Batch Start:  164000\n",
      "Batch Start:  165000\n",
      "Batch Start:  166000\n",
      "Batch Start:  167000\n",
      "Batch Start:  168000\n",
      "Batch Start:  169000\n",
      "Batch Start:  170000\n",
      "Batch Start:  171000\n",
      "Batch Start:  172000\n",
      "Batch Start:  173000\n",
      "Batch Start:  174000\n",
      "Batch Start:  175000\n",
      "Batch Start:  176000\n",
      "Batch Start:  177000\n",
      "Batch Start:  178000\n",
      "Batch Start:  179000\n",
      "Batch Start:  180000\n",
      "Batch Start:  181000\n",
      "Batch Start:  182000\n",
      "Batch Start:  183000\n",
      "Batch Start:  184000\n",
      "Batch Start:  185000\n",
      "Batch Start:  186000\n",
      "Batch Start:  187000\n",
      "Batch Start:  188000\n",
      "Batch Start:  189000\n",
      "             TCR_name      TRBV     TRBJ           TRB_CDR3   TRBC  \\\n",
      "0  AAACCTGAGACAAAGG-4  TRBV10-3  TRBJ2-1  CAISDPGLAGGGGEQFF  TRBC2   \n",
      "1  AAACCTGAGACAAAGG-4  TRBV10-3  TRBJ2-1  CAISDPGLAGGGGEQFF  TRBC2   \n",
      "2  AAACCTGAGACAAAGG-4  TRBV10-3  TRBJ2-1  CAISDPGLAGGGGEQFF  TRBC2   \n",
      "3  AAACCTGAGACAAAGG-4  TRBV10-3  TRBJ2-1  CAISDPGLAGGGGEQFF  TRBC2   \n",
      "4  AAACCTGAGACAAAGG-4  TRBV10-3  TRBJ2-1  CAISDPGLAGGGGEQFF  TRBC2   \n",
      "\n",
      "      Epitope          MHC Binding task  \n",
      "0   KTWGQYWQV  HLA-A*02:01       0  nan  \n",
      "1  ELAGIGILTV  HLA-A*02:01       0  nan  \n",
      "2  CLLWSFQTSA  HLA-A*02:01       0  nan  \n",
      "3   IMDQVPFSV  HLA-A*02:01       0  nan  \n",
      "4   SLLMWITQV  HLA-A*02:01       0  nan  \n"
     ]
    }
   ],
   "source": [
    "#Dieser Code für ganzen Datensatz laufen lassen\n",
    "import re\n",
    "import pandas as pd\n",
    "\n",
    "# Annahme: all_donors_consensus und all_donors_meta sind bereits geladen und gefiltert\n",
    "\n",
    "# Festlegen der Batch-Größe für die Verarbeitung\n",
    "batch_size = 1000  # Passe diese Zahl je nach Speicherressourcen an\n",
    "\n",
    "# Identifizieren von Epitope-Spalten, aber ohne \"NR(B0801)_AAKGRGAAL_NC_binder\"\n",
    "epitope_columns = [col for col in all_donors_consensus.columns if 'Cancer_binder' in col and col != \"NR(B0801)_AAKGRGAAL_NC_binder\"]\n",
    "\n",
    "# Liste für alle Batch-Ergebnisse\n",
    "all_batches = []\n",
    "\n",
    "# Verarbeite `all_donors_consensus` in Batches\n",
    "for batch_start in range(0, len(all_donors_consensus), batch_size):\n",
    "    print(\"Batch Start: \", batch_start)\n",
    "    # Batch definieren\n",
    "    batch = all_donors_consensus.iloc[batch_start:batch_start + batch_size]\n",
    "    \n",
    "    # Filtern auf Zeilen, die 'TRB:' enthalten\n",
    "    batch_trb = batch[batch['cell_clono_cdr3_aa'].str.contains(\"TRB:\", na=False)]\n",
    "\n",
    "    # Liste, um Zeilen für diesen Batch zu speichern\n",
    "    expanded_rows = []\n",
    "    \n",
    "    # Iteriere durch jede Zeile im Batch\n",
    "    for _, row in batch_trb.iterrows():\n",
    "        for col in epitope_columns:\n",
    "            # Extrahiere MHC und Epitope\n",
    "            match = re.match(r'([A-Z0-9]+)_([A-Z]+)_.*_binder', col)\n",
    "            if match:\n",
    "                mhc_raw, epitope = match.groups()\n",
    "                mhc_formatted = f'HLA-{mhc_raw[0]}*{mhc_raw[1:3]}:{mhc_raw[3:]}'\n",
    "\n",
    "                # Füge `Epitope` und `MHC` zur Zeile hinzu\n",
    "                new_row = row.copy()\n",
    "                new_row['Epitope'] = epitope\n",
    "                new_row['MHC'] = mhc_formatted\n",
    "\n",
    "                # Füge neue Zeile zur Batch-Liste hinzu\n",
    "                expanded_rows.append(new_row)\n",
    "    \n",
    "    # Erstelle einen DataFrame aus dem Batch\n",
    "    batch_df = pd.DataFrame(expanded_rows)\n",
    "    all_batches.append(batch_df)  # Speichere den Batch in der Liste\n",
    "\n",
    "# Kombiniere alle Batch-Ergebnisse zu einem DataFrame\n",
    "expanded_df = pd.concat(all_batches, ignore_index=True)\n",
    "\n",
    "# Nur die TRB-Chain-Einträge in `all_donors_meta` beibehalten\n",
    "all_donors_meta_trb = all_donors_meta[all_donors_meta['chain'] == 'TRB']\n",
    "\n",
    "# Zusammenführen der beiden DataFrames basierend auf der 'barcode' Spalte\n",
    "merged_df = pd.merge(all_donors_meta_trb, expanded_df[['barcode', 'Epitope', 'MHC']], on='barcode', how='inner')\n",
    "\n",
    "# Spalten umbenennen und Format anpassen\n",
    "merged_df = merged_df.rename(columns={\n",
    "    'barcode': 'TCR_name',\n",
    "    'v_gene': 'TRBV',\n",
    "    'j_gene': 'TRBJ',\n",
    "    'c_gene': 'TRBC',\n",
    "    'cdr3': 'TRB_CDR3'\n",
    "})\n",
    "\n",
    "# Fehlende Spalten auffüllen\n",
    "desired_columns = ['TCR_name', 'TRBV', 'TRBJ', 'TRB_CDR3', 'TRBC', 'Epitope', 'MHC', 'Binding', 'task']\n",
    "for col in desired_columns:\n",
    "    if col not in merged_df.columns:\n",
    "        merged_df[col] = 'nan' if col == 'task' else '0'\n",
    "\n",
    "# Nur die gewünschten Spalten beibehalten und Zeilen mit `None` in `TRB_CDR3` entfernen\n",
    "final_df = merged_df[desired_columns]\n",
    "final_df = final_df[final_df['TRB_CDR3'] != 'None']\n",
    "\n",
    "final_df = final_df[final_df['TRB_CDR3'].notna() & (final_df['TRB_CDR3'] != '')]\n",
    "\n",
    "# Ausgabe des ersten Teils des Ergebnisses zur Überprüfung\n",
    "print(final_df.head())\n",
    "\n",
    "# Speichern des kombinierten DataFrames\n",
    "output_path = '../data_10x/combined_output_with_epitope_mhc_TRB_only_expanded-gene.csv'\n",
    "final_df.to_csv(output_path, index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Datensätze wurden erfolgreich kombiniert und gespeichert.\n"
     ]
    }
   ],
   "source": [
    "# Mit bestehendem Train, Validation und Test File zusammenführen\n",
    "import pandas as pd\n",
    "from sklearn.model_selection import train_test_split\n",
    "\n",
    "# Lade den `combined_output_with_epitope_mhc_TRB_only_expanded.csv`-Datensatz\n",
    "beta_10x = pd.read_csv(\"../data_10x/combined_output_with_epitope_mhc_TRB_only_expanded.csv\", sep=',')\n",
    "\n",
    "# Schritt 1: Splitte `beta_10x` in train, validation und test\n",
    "train_split, test_split = train_test_split(beta_10x, test_size=0.3, random_state=42)\n",
    "validation_split, test_split = train_test_split(test_split, test_size=0.5, random_state=42)\n",
    "\n",
    "# Schritt 2: Lade die existierenden positiven Daten aus train, validation und test Dateien\n",
    "train_preneg = pd.read_csv(f\"./data_10x/splitted_datasets/{precision}/beta/train_prenegsamples.tsv\", sep='\\t')\n",
    "validation_preneg = pd.read_csv(f\"./data_10x/splitted_datasets/{precision}/beta/validation_prenegsamples.tsv\", sep='\\t')\n",
    "test_preneg = pd.read_csv(f\"./data_10x/splitted_datasets/{precision}/beta/test_prenegsamples.tsv\", sep='\\t')\n",
    "\n",
    "# Anzahl positiver Samples in train_split, validation_split und test_split\n",
    "num_train_pos = len(train_preneg)\n",
    "num_validation_pos = len(validation_preneg)\n",
    "num_test_pos = len(test_preneg)\n",
    "\n",
    "# Schritt 3: Berechne die Anzahl negativer Samples, um das Verhältnis zu erreichen\n",
    "# Training set: 1:1 ratio\n",
    "train_neg_needed = num_train_pos\n",
    "train_neg_sampled = train_split.sample(n=train_neg_needed, random_state=42, replace=False)\n",
    "\n",
    "# Validation set: 1:5 ratio\n",
    "validation_neg_needed = 5* num_validation_pos\n",
    "validation_neg_sampled = validation_split.sample(n=validation_neg_needed, random_state=42, replace=False)\n",
    "\n",
    "# Test set: 1:5 ratio\n",
    "test_neg_needed = 5* num_test_pos\n",
    "test_neg_sampled = test_split.sample(n=test_neg_needed, random_state=42, replace=False)\n",
    "\n",
    "# Schritt 4: Kombiniere die negativen und positiven Daten für train, validation und test\n",
    "train_combined = pd.concat([train_preneg, train_neg_sampled], ignore_index=True)\n",
    "validation_combined = pd.concat([validation_preneg, validation_neg_sampled], ignore_index=True)\n",
    "test_combined = pd.concat([test_preneg, test_neg_sampled], ignore_index=True)\n",
    "\n",
    "# Schritt 5: Speichern der kombinierten Datensätze\n",
    "output_dir = f'./data_10x/splitted_datasets/{precision}/beta/'\n",
    "train_combined.to_csv(output_dir + \"train.tsv\", sep='\\t', index=False)\n",
    "validation_combined.to_csv(output_dir + \"validation.tsv\", sep='\\t', index=False)\n",
    "test_combined.to_csv(output_dir + \"test.tsv\", sep='\\t', index=False)\n",
    "\n",
    "print(\"Datensätze wurden erfolgreich kombiniert und gespeichert.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TRB_CDR3 value types:\n",
      "TRB_CDR3\n",
      "<class 'str'>    2126313\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(\"TRB_CDR3 value types:\")\n",
    "print(final_df[\"TRB_CDR3\"].apply(type).value_counts())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Paired"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Batch Start:  0\n",
      "Batch Start:  1000\n",
      "Batch Start:  2000\n",
      "Batch Start:  3000\n",
      "Batch Start:  4000\n",
      "Batch Start:  5000\n",
      "Batch Start:  6000\n",
      "Batch Start:  7000\n",
      "Batch Start:  8000\n",
      "Batch Start:  9000\n",
      "Batch Start:  10000\n",
      "Batch Start:  11000\n",
      "Batch Start:  12000\n",
      "Batch Start:  13000\n",
      "Batch Start:  14000\n",
      "Batch Start:  15000\n",
      "Batch Start:  16000\n",
      "Batch Start:  17000\n",
      "Batch Start:  18000\n",
      "Batch Start:  19000\n",
      "Batch Start:  20000\n",
      "Batch Start:  21000\n",
      "Batch Start:  22000\n",
      "Batch Start:  23000\n",
      "Batch Start:  24000\n",
      "Batch Start:  25000\n",
      "Batch Start:  26000\n",
      "Batch Start:  27000\n",
      "Batch Start:  28000\n",
      "Batch Start:  29000\n",
      "Batch Start:  30000\n",
      "Batch Start:  31000\n",
      "Batch Start:  32000\n",
      "Batch Start:  33000\n",
      "Batch Start:  34000\n",
      "Batch Start:  35000\n",
      "Batch Start:  36000\n",
      "Batch Start:  37000\n",
      "Batch Start:  38000\n",
      "Batch Start:  39000\n",
      "Batch Start:  40000\n",
      "Batch Start:  41000\n",
      "Batch Start:  42000\n",
      "Batch Start:  43000\n",
      "Batch Start:  44000\n",
      "Batch Start:  45000\n",
      "Batch Start:  46000\n",
      "Batch Start:  47000\n",
      "Batch Start:  48000\n",
      "Batch Start:  49000\n",
      "Batch Start:  50000\n",
      "Batch Start:  51000\n",
      "Batch Start:  52000\n",
      "Batch Start:  53000\n",
      "Batch Start:  54000\n",
      "Batch Start:  55000\n",
      "Batch Start:  56000\n",
      "Batch Start:  57000\n",
      "Batch Start:  58000\n",
      "Batch Start:  59000\n",
      "Batch Start:  60000\n",
      "Batch Start:  61000\n",
      "Batch Start:  62000\n",
      "Batch Start:  63000\n",
      "Batch Start:  64000\n",
      "Batch Start:  65000\n",
      "Batch Start:  66000\n",
      "Batch Start:  67000\n",
      "Batch Start:  68000\n",
      "Batch Start:  69000\n",
      "Batch Start:  70000\n",
      "Batch Start:  71000\n",
      "Batch Start:  72000\n",
      "Batch Start:  73000\n",
      "Batch Start:  74000\n",
      "Batch Start:  75000\n",
      "Batch Start:  76000\n",
      "Batch Start:  77000\n",
      "Batch Start:  78000\n",
      "Batch Start:  79000\n",
      "Batch Start:  80000\n",
      "Batch Start:  81000\n",
      "Batch Start:  82000\n",
      "Batch Start:  83000\n",
      "Batch Start:  84000\n",
      "Batch Start:  85000\n",
      "Batch Start:  86000\n",
      "Batch Start:  87000\n",
      "Batch Start:  88000\n",
      "Batch Start:  89000\n",
      "Batch Start:  90000\n",
      "Batch Start:  91000\n",
      "Batch Start:  92000\n",
      "Batch Start:  93000\n",
      "Batch Start:  94000\n",
      "Batch Start:  95000\n",
      "Batch Start:  96000\n",
      "Batch Start:  97000\n",
      "Batch Start:  98000\n",
      "Batch Start:  99000\n",
      "Batch Start:  100000\n",
      "Batch Start:  101000\n",
      "Batch Start:  102000\n",
      "Batch Start:  103000\n",
      "Batch Start:  104000\n",
      "Batch Start:  105000\n",
      "Batch Start:  106000\n",
      "Batch Start:  107000\n",
      "Batch Start:  108000\n",
      "Batch Start:  109000\n",
      "Batch Start:  110000\n",
      "Batch Start:  111000\n",
      "Batch Start:  112000\n",
      "Batch Start:  113000\n",
      "Batch Start:  114000\n",
      "Batch Start:  115000\n",
      "Batch Start:  116000\n",
      "Batch Start:  117000\n",
      "Batch Start:  118000\n",
      "Batch Start:  119000\n",
      "Batch Start:  120000\n",
      "Batch Start:  121000\n",
      "Batch Start:  122000\n",
      "Batch Start:  123000\n",
      "Batch Start:  124000\n",
      "Batch Start:  125000\n",
      "Batch Start:  126000\n",
      "Batch Start:  127000\n",
      "Batch Start:  128000\n",
      "Batch Start:  129000\n",
      "Batch Start:  130000\n",
      "Batch Start:  131000\n",
      "Batch Start:  132000\n",
      "Batch Start:  133000\n",
      "Batch Start:  134000\n",
      "Batch Start:  135000\n",
      "Batch Start:  136000\n",
      "Batch Start:  137000\n",
      "Batch Start:  138000\n",
      "Batch Start:  139000\n",
      "Batch Start:  140000\n",
      "Batch Start:  141000\n",
      "Batch Start:  142000\n",
      "Batch Start:  143000\n",
      "Batch Start:  144000\n",
      "Batch Start:  145000\n",
      "Batch Start:  146000\n",
      "Batch Start:  147000\n",
      "Batch Start:  148000\n",
      "Batch Start:  149000\n",
      "Batch Start:  150000\n",
      "Batch Start:  151000\n",
      "Batch Start:  152000\n",
      "Batch Start:  153000\n",
      "Batch Start:  154000\n",
      "Batch Start:  155000\n",
      "Batch Start:  156000\n",
      "Batch Start:  157000\n",
      "Batch Start:  158000\n",
      "Batch Start:  159000\n",
      "Batch Start:  160000\n",
      "Batch Start:  161000\n",
      "Batch Start:  162000\n",
      "Batch Start:  163000\n",
      "Batch Start:  164000\n",
      "Batch Start:  165000\n",
      "Batch Start:  166000\n",
      "Batch Start:  167000\n",
      "Batch Start:  168000\n",
      "Batch Start:  169000\n",
      "Batch Start:  170000\n",
      "Batch Start:  171000\n",
      "Batch Start:  172000\n",
      "Batch Start:  173000\n",
      "Batch Start:  174000\n",
      "Batch Start:  175000\n",
      "Batch Start:  176000\n",
      "Batch Start:  177000\n",
      "Batch Start:  178000\n",
      "Batch Start:  179000\n",
      "Batch Start:  180000\n",
      "Batch Start:  181000\n",
      "Batch Start:  182000\n",
      "Batch Start:  183000\n",
      "Batch Start:  184000\n",
      "Batch Start:  185000\n",
      "Batch Start:  186000\n",
      "Batch Start:  187000\n",
      "Batch Start:  188000\n",
      "Batch Start:  189000\n",
      "             TCR_name       TRAV    TRAJ           TRA_CDR3      TRBV  \\\n",
      "0  AAACCTGAGACAAAGG-4  TRAV29DV5  TRAJ44  CAASVSIWTGTASKLTF  TRBV10-3   \n",
      "1  AAACCTGAGACAAAGG-4  TRAV29DV5  TRAJ44  CAASVSIWTGTASKLTF  TRBV10-3   \n",
      "2  AAACCTGAGACAAAGG-4  TRAV29DV5  TRAJ44  CAASVSIWTGTASKLTF  TRBV10-3   \n",
      "3  AAACCTGAGACAAAGG-4  TRAV29DV5  TRAJ44  CAASVSIWTGTASKLTF  TRBV10-3   \n",
      "4  AAACCTGAGACAAAGG-4  TRAV29DV5  TRAJ44  CAASVSIWTGTASKLTF  TRBV10-3   \n",
      "\n",
      "      TRBJ           TRB_CDR3  TRAC   TRBC     Epitope          MHC Binding  \\\n",
      "0  TRBJ2-1  CAISDPGLAGGGGEQFF  TRAC  TRBC2   KTWGQYWQV  HLA-A*02:01       0   \n",
      "1  TRBJ2-1  CAISDPGLAGGGGEQFF  TRAC  TRBC2  ELAGIGILTV  HLA-A*02:01       0   \n",
      "2  TRBJ2-1  CAISDPGLAGGGGEQFF  TRAC  TRBC2  CLLWSFQTSA  HLA-A*02:01       0   \n",
      "3  TRBJ2-1  CAISDPGLAGGGGEQFF  TRAC  TRBC2   IMDQVPFSV  HLA-A*02:01       0   \n",
      "4  TRBJ2-1  CAISDPGLAGGGGEQFF  TRAC  TRBC2   SLLMWITQV  HLA-A*02:01       0   \n",
      "\n",
      "  task  \n",
      "0  nan  \n",
      "1  nan  \n",
      "2  nan  \n",
      "3  nan  \n",
      "4  nan  \n"
     ]
    }
   ],
   "source": [
    "#Dieser Code für ganzen Datensatz laufen lassen\n",
    "import re\n",
    "import pandas as pd\n",
    "\n",
    "# Annahme: all_donors_consensus und all_donors_meta sind bereits geladen und gefiltert\n",
    "\n",
    "# Festlegen der Batch-Größe für die Verarbeitung\n",
    "batch_size = 1000\n",
    "\n",
    "# Identifizieren von Epitope-Spalten, aber ohne \"NR(B0801)_AAKGRGAAL_NC_binder\"\n",
    "epitope_columns = [col for col in all_donors_consensus.columns if 'Cancer_binder' in col and col != \"NR(B0801)_AAKGRGAAL_NC_binder\"]\n",
    "\n",
    "# Liste für alle Batch-Ergebnisse\n",
    "all_batches = []\n",
    "\n",
    "# Verarbeite `all_donors_consensus` in Batches\n",
    "for batch_start in range(0, len(all_donors_consensus), batch_size):\n",
    "    print(\"Batch Start: \", batch_start)\n",
    "    # Definiere Batch\n",
    "    batch = all_donors_consensus.iloc[batch_start:batch_start + batch_size]\n",
    "    \n",
    "    # Filtern auf Zeilen, die sowohl 'TRA:' als auch 'TRB:' in 'cell_clono_cdr3_aa' enthalten\n",
    "    batch_paired = batch[\n",
    "        batch['cell_clono_cdr3_aa'].str.contains(\"TRA:\", na=False) &\n",
    "        batch['cell_clono_cdr3_aa'].str.contains(\"TRB:\", na=False)\n",
    "    ]\n",
    "\n",
    "    # Liste, um Zeilen für diesen Batch zu speichern\n",
    "    expanded_rows = []\n",
    "    \n",
    "    # Iteriere durch jede Zeile im Batch\n",
    "    for _, row in batch_paired.iterrows():\n",
    "        for col in epitope_columns:\n",
    "            # Extrahiere MHC und Epitope\n",
    "            match = re.match(r'([A-Z0-9]+)_([A-Z]+)_.*_binder', col)\n",
    "            if match:\n",
    "                mhc_raw, epitope = match.groups()\n",
    "                mhc_formatted = f'HLA-{mhc_raw[0]}*{mhc_raw[1:3]}:{mhc_raw[3:]}'\n",
    "\n",
    "                # Füge `Epitope` und `MHC` zur Zeile hinzu\n",
    "                new_row = row.copy()\n",
    "                new_row['Epitope'] = epitope\n",
    "                new_row['MHC'] = mhc_formatted\n",
    "\n",
    "                # Neue Zeile zur Batch-Liste hinzufügen\n",
    "                expanded_rows.append(new_row)\n",
    "\n",
    "    # Erstelle einen DataFrame aus dem Batch und füge ihn zur Gesamtliste hinzu\n",
    "    batch_df = pd.DataFrame(expanded_rows)\n",
    "    all_batches.append(batch_df)\n",
    "\n",
    "# Kombiniere alle Batch-Ergebnisse zu einem DataFrame\n",
    "expanded_df = pd.concat(all_batches, ignore_index=True)\n",
    "\n",
    "# Nur die Paired-Einträge in `all_donors_meta` beibehalten\n",
    "# Filtern auf Barcodes, die sowohl eine TRA- als auch eine TRB-Kette haben\n",
    "paired_barcodes = all_donors_meta.groupby('barcode').filter(\n",
    "    lambda x: set(x['chain']) == {'TRA', 'TRB'}\n",
    ")['barcode'].unique()\n",
    "all_donors_meta_paired = all_donors_meta[all_donors_meta['barcode'].isin(paired_barcodes)]\n",
    "\n",
    "# Split `all_donors_meta_paired` nach `chain` in separate DataFrames für TRA und TRB\n",
    "alpha_chain = all_donors_meta_paired[all_donors_meta_paired['chain'] == 'TRA'].rename(\n",
    "    columns={'v_gene': 'TRAV', 'j_gene': 'TRAJ', 'cdr3': 'TRA_CDR3', 'c_gene': 'TRAC'}\n",
    ")\n",
    "beta_chain = all_donors_meta_paired[all_donors_meta_paired['chain'] == 'TRB'].rename(\n",
    "    columns={'v_gene': 'TRBV', 'j_gene': 'TRBJ', 'cdr3': 'TRB_CDR3', 'c_gene': 'TRBC'}\n",
    ")\n",
    "\n",
    "# Zusammenführen von alpha_chain und beta_chain anhand der gemeinsamen 'barcode'-Spalte\n",
    "paired_meta = pd.merge(alpha_chain, beta_chain, on='barcode', suffixes=('_alpha', '_beta'))\n",
    "\n",
    "# Zusammenführen von `paired_meta` mit `expanded_df` anhand der 'barcode'-Spalte\n",
    "merged_df = pd.merge(paired_meta, expanded_df[['barcode', 'Epitope', 'MHC']], on='barcode', how='inner')\n",
    "\n",
    "# Spalten umbenennen und Format anpassen\n",
    "merged_df = merged_df.rename(columns={'barcode': 'TCR_name'})\n",
    "\n",
    "# Fehlende Spalten auffüllen\n",
    "desired_columns = [\n",
    "    'TCR_name', 'TRAV', 'TRAJ', 'TRA_CDR3', 'TRBV', 'TRBJ', 'TRB_CDR3', 'TRAC', 'TRBC', \n",
    "    'Epitope', 'MHC', 'Binding', 'task'\n",
    "]\n",
    "for col in desired_columns:\n",
    "    if col not in merged_df.columns:\n",
    "        merged_df[col] = 'nan' if col == 'task' else '0'\n",
    "\n",
    "# Nur die gewünschten Spalten beibehalten und Zeilen mit `None` in `TRB_CDR3` entfernen\n",
    "final_df = merged_df[desired_columns]\n",
    "final_df = final_df[final_df['TRB_CDR3'] != 'None']\n",
    "\n",
    "final_df = final_df[\n",
    "    final_df['TRB_CDR3'].notna() & (final_df['TRB_CDR3'] != '') &\n",
    "    final_df['TRA_CDR3'].notna() & (final_df['TRA_CDR3'] != '')\n",
    "]\n",
    "\n",
    "# Ausgabe des ersten Teils des Ergebnisses zur Überprüfung\n",
    "print(final_df.head())\n",
    "\n",
    "# Optional: Speichern des kombinierten DataFrames\n",
    "output_path = '../data_10x/combined_output_with_epitope_mhc_paired_only_expanded-gene.csv'\n",
    "final_df.to_csv(output_path, index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Datensätze wurden erfolgreich kombiniert und gespeichert.\n"
     ]
    }
   ],
   "source": [
    "# Mit bestehendem Train, Validation und Test File zusammenführen\n",
    "import pandas as pd\n",
    "from sklearn.model_selection import train_test_split\n",
    "\n",
    "# Lade den `combined_output_with_epitope_mhc_TRB_only_expanded.csv`-Datensatz\n",
    "beta_10x = pd.read_csv(\"../data_10x/combined_output_with_epitope_mhc_paired_only_expanded.csv\", sep=',')\n",
    "\n",
    "# Schritt 1: Splitte negativen Samples `beta_10x` in train, validation und test\n",
    "train_split, test_split = train_test_split(beta_10x, test_size=0.3, random_state=42)\n",
    "validation_split, test_split = train_test_split(test_split, test_size=0.5, random_state=42)\n",
    "\n",
    "# Schritt 2: Lade die existierenden positiven Samples aus train, validation und test Dateien\n",
    "train_preneg = pd.read_csv(f\"./data_10x/splitted_datasets/{precision}/paired/train_prenegsamples.tsv\", sep='\\t')\n",
    "validation_preneg = pd.read_csv(f\"./data_10x/splitted_datasets/{precision}/paired/validation_prenegsamples.tsv\", sep='\\t')\n",
    "test_preneg = pd.read_csv(f\"./data_10x/splitted_datasets/{precision}/paired/test_prenegsamples.tsv\", sep='\\t')\n",
    "\n",
    "# Anzahl positiver Samples in train_split, validation_split und test_split\n",
    "num_train_pos = len(train_preneg)\n",
    "num_validation_pos = len(validation_preneg)\n",
    "num_test_pos = len(test_preneg)\n",
    "\n",
    "# Schritt 3: Berechne die Anzahl negativer Samples, um das Verhältnis zu erreichen\n",
    "# Training set: 1:1 ratio\n",
    "train_neg_needed = num_train_pos\n",
    "train_neg_sampled = train_split.sample(n=train_neg_needed, random_state=42, replace=False)\n",
    "\n",
    "# Validation set: 1:5 ratio\n",
    "validation_neg_needed = 5* num_validation_pos\n",
    "validation_neg_sampled = validation_split.sample(n=validation_neg_needed, random_state=42, replace=False)\n",
    "\n",
    "# Test set: 1:5 ratio\n",
    "test_neg_needed = 5* num_test_pos\n",
    "test_neg_sampled = test_split.sample(n=test_neg_needed, random_state=42, replace=False)\n",
    "\n",
    "# Schritt 4: Kombiniere die negativen und positiven Daten für train, validation und test\n",
    "train_combined = pd.concat([train_preneg, train_neg_sampled], ignore_index=True)\n",
    "validation_combined = pd.concat([validation_preneg, validation_neg_sampled], ignore_index=True)\n",
    "test_combined = pd.concat([test_preneg, test_neg_sampled], ignore_index=True)\n",
    "\n",
    "# Schritt 5: Speichern der kombinierten Datensätze\n",
    "output_dir = f'./data_10x/splitted_datasets/{precision}/paired/'\n",
    "train_combined.to_csv(output_dir + \"train.tsv\", sep='\\t', index=False)\n",
    "validation_combined.to_csv(output_dir + \"validation.tsv\", sep='\\t', index=False)\n",
    "test_combined.to_csv(output_dir + \"test.tsv\", sep='\\t', index=False)\n",
    "\n",
    "print(\"Datensätze wurden erfolgreich kombiniert und gespeichert.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TCR_name    object\n",
      "TRAV        object\n",
      "TRAJ        object\n",
      "TRA_CDR3    object\n",
      "TRBV        object\n",
      "TRBJ        object\n",
      "TRB_CDR3    object\n",
      "TRAC        object\n",
      "TRBC        object\n",
      "Epitope     object\n",
      "MHC         object\n",
      "Binding     object\n",
      "task        object\n",
      "dtype: object\n"
     ]
    }
   ],
   "source": [
    "print(final_df.dtypes)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TRA_CDR3 value types:\n",
      "TRA_CDR3\n",
      "<class 'str'>    2028672\n",
      "Name: count, dtype: int64\n",
      "TRB_CDR3 value types:\n",
      "TRB_CDR3\n",
      "<class 'str'>    2028672\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(\"TRA_CDR3 value types:\")\n",
    "print(final_df[\"TRA_CDR3\"].apply(type).value_counts())\n",
    "\n",
    "print(\"TRB_CDR3 value types:\")\n",
    "print(final_df[\"TRB_CDR3\"].apply(type).value_counts())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mix generierte negative Samples mit 10X 1:1 ratio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Datei-Paths 10X\n",
    "train_file_10x_path = f\"./splitted_datasets/{precision}/paired/train.tsv\"\n",
    "validation_file_10x_path = f\"./splitted_datasets/{precision}/paired/validation.tsv\"\n",
    "test_file_10x_path = f\"./splitted_datasets/{precision}/paired/test.tsv\"\n",
    "\n",
    "# Datei-Paths BA\n",
    "train_file_path = f'./../../BA/BA_ZHAW/data/splitted_datasets/allele/paired/train.tsv'\n",
    "validation_file_path = f'./../../BA/BA_ZHAW/data/splitted_datasets/allele/paired/validation.tsv'\n",
    "test_file_path = f'./../../BA/BA_ZHAW/data/splitted_datasets/allele/paired/test.tsv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task Classification >>> wichtig: bei paired_reclassification.ipynb _10x zu unterst hinzufügen bei main_project...\n",
    "The classification in the split notebook correct for positive only data. After adding negative data, some classifications might be wrong."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "paired_output_folder = f'{pipeline_data_splitted}/{precision}/paired'\n",
    "validation_file_name = 'validation.tsv'\n",
    "test_file_name = 'test.tsv'\n",
    "train_file_name = 'train.tsv'\n",
    "beta_output_folder = f'{pipeline_data_splitted}/{precision}/beta'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "# do the classification for paired data\n",
    "paired = True\n",
    "train_data_path = f'{paired_output_folder}/{train_file_name}'\n",
    "test_data_path = f'{paired_output_folder}/{test_file_name}'\n",
    "validation_data_path = f'{paired_output_folder}/{validation_file_name}'\n",
    "\n",
    "%run ./data_scripts/data_preparation/classification.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene\n",
      "./data_10x/splitted_datasets/gene/paired/train.tsv\n",
      "                  TCR_name        TRAV    TRAJ          TRA_CDR3      TRBV  \\\n",
      "0                    17414  TRAV29/DV5  TRAJ37    CAASALGNTGKLIF   TRBV4-1   \n",
      "1                    20147       TRAV5  TRAJ20     CAEIRANDYKLSF   TRBV5-1   \n",
      "2                    16532      TRAV17  TRAJ17  CAALDGIKAAGNKLTF    TRBV19   \n",
      "3                    21795         NaN     NaN      CAFLGGANNLFF       NaN   \n",
      "4                     8426      TRAV17   TRAJ7    CATGLYYGNNRLAF  TRBV11-2   \n",
      "...                    ...         ...     ...               ...       ...   \n",
      "43351   TAGAGCTCATTCCTCG-5      TRAV20  TRAJ52  CAVRRAGGTSYGKLTF   TRBV3-1   \n",
      "43352  TCTTCGGCAAACGTGG-40    TRAV38-1  TRAJ26   CAFNERDNYGQNFVF  TRBV10-1   \n",
      "43353  GTCACGGAGGTAGCCA-20     TRAV1-2   TRAJ4       CAVRNYNKLIF   TRBV4-3   \n",
      "43354  ACGTCAACATACTCTT-23     TRAV8-3   TRAJ7     CAVFLYGNNRLAF  TRBV29-1   \n",
      "43355  GAGTCCGAGGAGTAGA-25       TRAV5   TRAJ6     CAEMEGGSYIPTF   TRBV4-2   \n",
      "\n",
      "          TRBJ           TRB_CDR3  TRAC   TRBC    Epitope          MHC  \\\n",
      "0      TRBJ1-1     CASSPGAGDTEAFF   NaN    NaN  KLGGALQAK        HLA-A   \n",
      "1      TRBJ1-2        CASSLGQRYTF   NaN    NaN  KLGGALQAK        HLA-A   \n",
      "2      TRBJ1-1    CASKTGGALNTEAFF   NaN    NaN  KLGGALQAK        HLA-A   \n",
      "3          NaN     CASSLEWGGETQYF   NaN    NaN  KLGGALQAK        HLA-A   \n",
      "4      TRBJ2-2  CASSLGVAGTNTGELFF   NaN    NaN  KLGGALQAK        HLA-A   \n",
      "...        ...                ...   ...    ...        ...          ...   \n",
      "43351  TRBJ2-7   CASSQVSLTGGYEQYF  TRAC  TRBC2  KVAELVHFL  HLA-A*02:01   \n",
      "43352  TRBJ2-1     CASSRDSSANEQFF  TRAC  TRBC2  KVAELVHFL  HLA-A*02:01   \n",
      "43353  TRBJ2-1   CASSRSTLDSSNEQFF  TRAC  TRBC2  KVAELVHFL  HLA-A*02:01   \n",
      "43354  TRBJ2-7      CSVSASGGDEQYF  TRAC  TRBC2  SLLMWITQV  HLA-A*02:01   \n",
      "43355  TRBJ2-2    CASSQRPSEVGELFF  TRAC  TRBC2  RIAAWMATY  HLA-A*03:01   \n",
      "\n",
      "       Binding  task  \n",
      "0            1  TPP1  \n",
      "1            1  TPP1  \n",
      "2            1  TPP1  \n",
      "3            1  TPP1  \n",
      "4            1  TPP1  \n",
      "...        ...   ...  \n",
      "43351        0  TPP2  \n",
      "43352        0  TPP1  \n",
      "43353        0  TPP2  \n",
      "43354        0  TPP1  \n",
      "43355        0  TPP1  \n",
      "\n",
      "[43356 rows x 13 columns]\n",
      "train+validate data has 110766 entries\n",
      "test data has 43356 entries\n",
      "test data has 37667 TPP1 tasks (old value: 22545) (seen tcr & seen epitopes).\n",
      "test data has 5400 TPP2 tasks (old value: 20522) (unseen tcr & seen epitopes).\n",
      "test data has 227 TPP3 tasks (old value: 289) (unseen tcr & unseen epitope).\n",
      "test data has 62 TPP4 tasks (old value: 0) (seen tcr & unseen epitope).\n",
      "the train/test ratio is 0.7186903881340756/0.2813096118659244\n",
      "./data_10x/splitted_datasets/gene/paired/test_reclassified_paired_specific.tsv\n",
      "uploading dataset to dataset-gene_10x\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[34m\u001b[1mwandb\u001b[0m: Using wandb-core as the SDK backend.  Please refer to https://wandb.me/wandb-core for more information.\n",
      "\u001b[34m\u001b[1mwandb\u001b[0m: Currently logged in as: \u001b[33mfrohoari\u001b[0m (\u001b[33mpa_cancerimmunotherapy\u001b[0m). Use \u001b[1m`wandb login --relogin`\u001b[0m to force relogin\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "Tracking run with wandb version 0.18.7"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "Run data is saved locally in <code>/home/ubuntu/PA-Cancer-Immunotherapy-Transformer/BA_ZHAW/wandb/run-20241120_183936-ozsft79l</code>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "Syncing run <strong><a href='https://wandb.ai/pa_cancerimmunotherapy/dataset-gene_10x/runs/ozsft79l' target=\"_blank\">fast-lion-28</a></strong> to <a href='https://wandb.ai/pa_cancerimmunotherapy/dataset-gene_10x' target=\"_blank\">Weights & Biases</a> (<a href='https://wandb.me/developer-guide' target=\"_blank\">docs</a>)<br/>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       " View project at <a href='https://wandb.ai/pa_cancerimmunotherapy/dataset-gene_10x' target=\"_blank\">https://wandb.ai/pa_cancerimmunotherapy/dataset-gene_10x</a>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       " View run at <a href='https://wandb.ai/pa_cancerimmunotherapy/dataset-gene_10x/runs/ozsft79l' target=\"_blank\">https://wandb.ai/pa_cancerimmunotherapy/dataset-gene_10x/runs/ozsft79l</a>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[34m\u001b[1mwandb\u001b[0m: Adding directory to artifact (./data_10x/splitted_datasets/gene/paired)... Done. 0.5s\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6da945f1d6ef4d7eb4da1e5c3166f09c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "VBox(children=(Label(value='1.338 MB of 19.980 MB uploaded\\r'), FloatProgress(value=0.06694487096815765, max=1…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       " View run <strong style=\"color:#cdcd00\">fast-lion-28</strong> at: <a href='https://wandb.ai/pa_cancerimmunotherapy/dataset-gene_10x/runs/ozsft79l' target=\"_blank\">https://wandb.ai/pa_cancerimmunotherapy/dataset-gene_10x/runs/ozsft79l</a><br/> View project at: <a href='https://wandb.ai/pa_cancerimmunotherapy/dataset-gene_10x' target=\"_blank\">https://wandb.ai/pa_cancerimmunotherapy/dataset-gene_10x</a><br/>Synced 5 W&B file(s), 0 media file(s), 5 artifact file(s) and 0 other file(s)"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "Find logs at: <code>./wandb/run-20241120_183936-ozsft79l/logs</code>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# extended classification for paired data\n",
    "test_path = f'{paired_output_folder}/{test_file_name}'\n",
    "train_path = f'{paired_output_folder}/{train_file_name}'\n",
    "validation_path = f'{paired_output_folder}/{validation_file_name}'\n",
    "output_path = f'{paired_output_folder}/test_reclassified_paired_specific.tsv'\n",
    "paired_data_path = paired_output_folder\n",
    "alpha_cdr3_name = 'TRA_CDR3'\n",
    "beta_cdr3_name = 'TRB_CDR3'\n",
    "epitope_name = 'Epitope'\n",
    "task_name = 'task'\n",
    "\n",
    "%run ./data_scripts/data_preparation/paired_reclassification.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_53158/1815672389.py:1: DtypeWarning: Columns (0,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df_train = pd.read_csv(train_data_path, sep=\"\\t\")\n",
      "/tmp/ipykernel_53158/1815672389.py:2: DtypeWarning: Columns (8) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df_test = pd.read_csv(test_data_path, sep=\"\\t\")\n"
     ]
    }
   ],
   "source": [
    "# do the classification for beta data\n",
    "paired = False\n",
    "train_data_path = f'{beta_output_folder}/{train_file_name}'\n",
    "test_data_path = f'{beta_output_folder}/{test_file_name}'\n",
    "validation_data_path = f'{beta_output_folder}/{validation_file_name}'\n",
    "\n",
    "%run ./data_scripts/data_preparation/classification.ipynb"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the next two cells the classification is checked. If the output says \"Classification is correct\", everything is fine."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "train+validate data has 110766 entries\n",
      "test data has 43356 entries\n",
      "test data has 22545 TPP1 tasks (seen tcr & seen epitopes).\n",
      "test data has 20522 TPP2 tasks (unseen tcr & seen epitopes).\n",
      "test data has 289 TPP3 tasks (unseen tcr & unseen epitope).\n",
      "test data has 0 TPP4 tasks (seen tcr & unseen epitope).\n",
      "the train/test ratio is 0.7186903881340756/0.2813096118659244\n",
      "Incorrectly set tasks:\n",
      "          Epitope TRA_CDR3           TRB_CDR3  task\n",
      "7228    SLLMWITQV      NaN  CASSLDLSGGITDTQYF  TPP2\n",
      "7248   YLNDHLEPWI      NaN     CSARDGRREYEQYF  TPP2\n",
      "7275    RIAAWMATY      NaN      CASSIRSSYEQYF  TPP2\n",
      "7281   CLLWSFQTSA      NaN     CASSSRTGYDGYTF  TPP2\n",
      "7292   ELAGIGILTV      NaN   CASSPGARHLEETQYF  TPP2\n",
      "...           ...      ...                ...   ...\n",
      "43291   KTWGQYWQV      NaN      CSVDLEANYGYTF  TPP2\n",
      "43316   KVLEYVIKV      NaN        CASRGAGELFF  TPP2\n",
      "43331   RIAAWMATY      NaN     CASSFSGNTGELFF  TPP2\n",
      "43338   SLLMWITQV      NaN    CASSQRPSEVGELFF  TPP2\n",
      "43345  YLNDHLEPWI      NaN   CASSPGGLAGADTQYF  TPP2\n",
      "\n",
      "[4385 rows x 4 columns]\n",
      "Correctness summary:\n",
      "is_correct\n",
      "True     38971\n",
      "False     4385\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# check task classification paired\n",
    "splitted_data_path = paired_output_folder\n",
    "\n",
    "%run ./data_scripts/data_preparation/check_task_classification_paired.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_53158/1838351243.py:18: DtypeWarning: Columns (0,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df_train = pd.read_csv(f\"{splitted_data_path}/{train_file_name}\", sep=\"\\t\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "train data has 251750 entries\n",
      "test data has 161844 entries\n",
      "test data has 140580 TPP1 tasks (seen tcr & seen epitopes).\n",
      "test data has 20961 TPP2 tasks (unseen tcr & seen epitopes).\n",
      "test data has 298 TPP3 tasks (unseen tcr & unseen epitope).\n",
      "test data has 5 TPP4 tasks (seen tcr & unseen epitope).\n",
      "the train/test ratio is 0.7187434831570021/0.28125651684299796\n",
      "Classification is correct.\n",
      "Correctness summary:\n",
      "is_correct\n",
      "True    161844\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# check task classification beta\n",
    "splitted_data_path = beta_output_folder\n",
    "\n",
    "%run ./data_scripts/data_preparation/check_task_classification_beta.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Datei-Paths\n",
    "train_file_path = f\"./splitted_datasets/{precision}/paired/train.tsv\"\n",
    "validation_file_path = f\"./splitted_datasets/{precision}/paired/validation.tsv\"\n",
    "test_file_path = f\"./splitted_datasets/{precision}/paired/test.tsv\"\n",
    "\n",
    "# Spalten, die überprüft und bereinigt werden sollen\n",
    "columns_to_check = [\"TCR_name\", \"TRA_CDR3\", \"TRB_CDR3\"]\n",
    "\n",
    "# Funktion zur Bereinigung von NaN-Werten\n",
    "def remove_nan_values(file_path, columns):\n",
    "    print(f\"Cleaning file: {file_path}\")\n",
    "    df = pd.read_csv(file_path, sep='\\t')\n",
    "    \n",
    "    # Entferne Zeilen mit NaN-Werten in den angegebenen Spalten\n",
    "    for column in columns:\n",
    "        if column in df.columns:\n",
    "            original_length = len(df)\n",
    "            df = df[df[column].notna()]\n",
    "            cleaned_length = len(df)\n",
    "            print(f\"Column '{column}' - Removed {original_length - cleaned_length} rows with NaN values\")\n",
    "        else:\n",
    "            print(f\"Column {column} not found in {file_path}\")\n",
    "    \n",
    "    # Speichere die bereinigte Datei\n",
    "    df.to_csv(file_path, sep='\\t', index=False)\n",
    "    print(f\"Cleaned file saved: {file_path}\")\n",
    "\n",
    "# Dateien bereinigen\n",
    "for file_path in [train_file_path, validation_file_path, test_file_path]:\n",
    "    remove_nan_values(file_path, columns_to_check)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Datei-Paths\n",
    "train_file_path = f\"./splitted_datasets/{precision}/paired/train.tsv\"\n",
    "validation_file_path = f\"./splitted_datasets/{precision}/paired/validation.tsv\"\n",
    "test_file_path = f\"./splitted_datasets/{precision}/paired/test.tsv\"\n",
    "\n",
    "# Spalten, die überprüft werden sollen\n",
    "columns_to_check = [\"TCR_name\", \"TRA_CDR3\", \"TRB_CDR3\"]\n",
    "\n",
    "# Funktion zur Identifikation von nicht-String-Werten\n",
    "def find_non_string_values(file_path, columns):\n",
    "    print(f\"Checking file: {file_path}\")\n",
    "    df = pd.read_csv(file_path, sep='\\t')\n",
    "    \n",
    "    for column in columns:\n",
    "        if column in df.columns:\n",
    "            non_string_values = df[~df[column].apply(lambda x: isinstance(x, str))][column]\n",
    "            print(f\"Column '{column}' - Non-string values found: {len(non_string_values)}\")\n",
    "            if len(non_string_values) > 0:\n",
    "                print(non_string_values)\n",
    "        else:\n",
    "            print(f\"Column {column} not found in {file_path}\")\n",
    "\n",
    "# Dateien überprüfen\n",
    "for file_path in [train_file_path, validation_file_path, test_file_path]:\n",
    "    find_non_string_values(file_path, columns_to_check)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Upload dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['.ipynb_checkpoints', 'validation_prenegsamples.tsv', 'test.tsv', 'train.tsv', 'validation.tsv', 'train_prenegsamples.tsv', 'test_prenegsamples.tsv']\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "print(os.listdir(path_to_data))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "uploading dataset to dataset-gene_10x\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "Tracking run with wandb version 0.18.7"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "Run data is saved locally in <code>/home/ubuntu/PA-Cancer-Immunotherapy-Transformer/BA_ZHAW/wandb/run-20241120_192146-gf6sdwqr</code>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "Syncing run <strong><a href='https://wandb.ai/pa_cancerimmunotherapy/dataset-gene_10x/runs/gf6sdwqr' target=\"_blank\">stoic-waterfall-32</a></strong> to <a href='https://wandb.ai/pa_cancerimmunotherapy/dataset-gene_10x' target=\"_blank\">Weights & Biases</a> (<a href='https://wandb.me/developer-guide' target=\"_blank\">docs</a>)<br/>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       " View project at <a href='https://wandb.ai/pa_cancerimmunotherapy/dataset-gene_10x' target=\"_blank\">https://wandb.ai/pa_cancerimmunotherapy/dataset-gene_10x</a>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       " View run at <a href='https://wandb.ai/pa_cancerimmunotherapy/dataset-gene_10x/runs/gf6sdwqr' target=\"_blank\">https://wandb.ai/pa_cancerimmunotherapy/dataset-gene_10x/runs/gf6sdwqr</a>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[34m\u001b[1mwandb\u001b[0m: Adding directory to artifact (./data_10x/splitted_datasets/gene/paired)... Done. 0.2s\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "05e28d007af743b881a1161f07842cd4",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "VBox(children=(Label(value='1.181 MB of 13.461 MB uploaded\\r'), FloatProgress(value=0.08776903208456056, max=1…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       " View run <strong style=\"color:#cdcd00\">stoic-waterfall-32</strong> at: <a href='https://wandb.ai/pa_cancerimmunotherapy/dataset-gene_10x/runs/gf6sdwqr' target=\"_blank\">https://wandb.ai/pa_cancerimmunotherapy/dataset-gene_10x/runs/gf6sdwqr</a><br/> View project at: <a href='https://wandb.ai/pa_cancerimmunotherapy/dataset-gene_10x' target=\"_blank\">https://wandb.ai/pa_cancerimmunotherapy/dataset-gene_10x</a><br/>Synced 5 W&B file(s), 0 media file(s), 4 artifact file(s) and 0 other file(s)"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "Find logs at: <code>./wandb/run-20241120_192146-gf6sdwqr/logs</code>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from dotenv import load_dotenv, find_dotenv\n",
    "load_dotenv()\n",
    "\n",
    "# upload paired data\n",
    "path_to_data = f'{pipeline_data_splitted}/{precision}/paired'\n",
    "dataset_name = f'paired_{precision}'\n",
    "#main_project_name = os.getenv(\"MAIN_PROJECT_NAME\")\n",
    "main_project_name = f\"dataset-{precision}_10x\"\n",
    "\n",
    "%run ./data_scripts/upload_datasets.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "uploading dataset to dataset-gene_10x\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "Tracking run with wandb version 0.18.7"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "Run data is saved locally in <code>/home/ubuntu/PA-Cancer-Immunotherapy-Transformer/BA_ZHAW/wandb/run-20241120_192155-cez4kzkf</code>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "Syncing run <strong><a href='https://wandb.ai/pa_cancerimmunotherapy/dataset-gene_10x/runs/cez4kzkf' target=\"_blank\">lucky-smoke-33</a></strong> to <a href='https://wandb.ai/pa_cancerimmunotherapy/dataset-gene_10x' target=\"_blank\">Weights & Biases</a> (<a href='https://wandb.me/developer-guide' target=\"_blank\">docs</a>)<br/>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       " View project at <a href='https://wandb.ai/pa_cancerimmunotherapy/dataset-gene_10x' target=\"_blank\">https://wandb.ai/pa_cancerimmunotherapy/dataset-gene_10x</a>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       " View run at <a href='https://wandb.ai/pa_cancerimmunotherapy/dataset-gene_10x/runs/cez4kzkf' target=\"_blank\">https://wandb.ai/pa_cancerimmunotherapy/dataset-gene_10x/runs/cez4kzkf</a>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[34m\u001b[1mwandb\u001b[0m: Adding directory to artifact (./data_10x/splitted_datasets/gene/beta)... Done. 0.1s\n"
     ]
    },
    {
     "data": {
      "text/html": [
       " View run <strong style=\"color:#cdcd00\">lucky-smoke-33</strong> at: <a href='https://wandb.ai/pa_cancerimmunotherapy/dataset-gene_10x/runs/cez4kzkf' target=\"_blank\">https://wandb.ai/pa_cancerimmunotherapy/dataset-gene_10x/runs/cez4kzkf</a><br/> View project at: <a href='https://wandb.ai/pa_cancerimmunotherapy/dataset-gene_10x' target=\"_blank\">https://wandb.ai/pa_cancerimmunotherapy/dataset-gene_10x</a><br/>Synced 5 W&B file(s), 0 media file(s), 0 artifact file(s) and 0 other file(s)"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "Find logs at: <code>./wandb/run-20241120_192155-cez4kzkf/logs</code>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# upload beta data\n",
    "path_to_data = f'{pipeline_data_splitted}/{precision}/beta'\n",
    "dataset_name = f'beta_{precision}'\n",
    "\n",
    "%run ./data_scripts/upload_datasets.ipynb"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create Embeddings >> vorher noch datacheck laufen lassen, um NAs Werte rauslöschen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "12.4\n"
     ]
    }
   ],
   "source": [
    "import torch\n",
    "print(torch.cuda.is_available())  # Sollte True zurückgeben\n",
    "print(torch.version.cuda)  # Sollte die richtige CUDA-Version anzeigen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using GPU: Tesla T4\n",
      "Loading: Rostlab/prot_t5_xl_half_uniref50-enc\n",
      "Model is on device: cuda:0\n",
      "Processing Batch:  0 64\n",
      "Processing Batch:  64 128\n",
      "Processing Batch:  128 192\n",
      "Processing Batch:  192 256\n",
      "Processing Batch:  256 320\n",
      "Processing Batch:  320 384\n",
      "Processing Batch:  384 448\n",
      "Processing Batch:  448 512\n",
      "Processing Batch:  512 576\n",
      "Processing Batch:  576 640\n",
      "Processing Batch:  640 704\n",
      "Processing Batch:  704 768\n",
      "Processing Batch:  768 832\n",
      "Processing Batch:  832 896\n",
      "Processing Batch:  896 960\n",
      "Processing Batch:  960 1024\n",
      "Processing Batch:  1024 1088\n",
      "Processing Batch:  1088 1152\n",
      "Processing Batch:  1152 1216\n",
      "Processing Batch:  1216 1280\n",
      "Processing Batch:  1280 1344\n",
      "Processing Batch:  1344 1408\n",
      "Processing Batch:  1408 1472\n",
      "Processing Batch:  1472 1536\n",
      "Processing Batch:  1536 1600\n",
      "Processing Batch:  1600 1664\n",
      "Processing Batch:  1664 1728\n",
      "Processing Batch:  1728 1792\n",
      "Processing Batch:  1792 1856\n",
      "Processing Batch:  1856 1920\n",
      "Processing Batch:  1920 1984\n",
      "Processing Batch:  1984 2048\n",
      "Processing Batch:  2048 2112\n",
      "Processing Batch:  2112 2176\n",
      "Processing Batch:  2176 2240\n",
      "Processing Batch:  2240 2304\n",
      "Processing Batch:  2304 2368\n",
      "Processing Batch:  2368 2432\n",
      "Processing Batch:  2432 2496\n",
      "Processing Batch:  2496 2560\n",
      "Processing Batch:  2560 2624\n",
      "Processing Batch:  2624 2688\n",
      "Processing Batch:  2688 2752\n",
      "Processing Batch:  2752 2816\n",
      "Processing Batch:  2816 2880\n",
      "Processing Batch:  2880 2944\n",
      "Processing Batch:  2944 3008\n",
      "Processing Batch:  3008 3072\n",
      "Processing Batch:  3072 3136\n",
      "Processing Batch:  3136 3200\n",
      "Processing Batch:  3200 3264\n",
      "Processing Batch:  3264 3328\n",
      "Processing Batch:  3328 3392\n",
      "Processing Batch:  3392 3456\n",
      "Processing Batch:  3456 3520\n",
      "Processing Batch:  3520 3584\n",
      "Processing Batch:  3584 3648\n",
      "Processing Batch:  3648 3712\n",
      "Processing Batch:  3712 3776\n",
      "Processing Batch:  3776 3840\n",
      "Processing Batch:  3840 3904\n",
      "Processing Batch:  3904 3968\n",
      "Processing Batch:  3968 4032\n",
      "Processing Batch:  4032 4096\n",
      "Processing Batch:  4096 4160\n",
      "Processing Batch:  4160 4224\n",
      "Processing Batch:  4224 4288\n",
      "Processing Batch:  4288 4352\n",
      "Processing Batch:  4352 4416\n",
      "Processing Batch:  4416 4480\n",
      "Processing Batch:  4480 4544\n",
      "Processing Batch:  4544 4608\n",
      "Processing Batch:  4608 4672\n",
      "Processing Batch:  4672 4736\n",
      "Processing Batch:  4736 4800\n",
      "Processing Batch:  4800 4864\n",
      "Processing Batch:  4864 4928\n",
      "Processing Batch:  4928 4992\n",
      "Processing Batch:  4992 5056\n",
      "Processing Batch:  5056 5120\n",
      "Processing Batch:  5120 5184\n",
      "Processing Batch:  5184 5248\n",
      "Processing Batch:  5248 5312\n",
      "Processing Batch:  5312 5376\n",
      "Processing Batch:  5376 5440\n",
      "Processing Batch:  5440 5504\n",
      "Processing Batch:  5504 5568\n",
      "Processing Batch:  5568 5632\n",
      "Processing Batch:  5632 5696\n",
      "Processing Batch:  5696 5760\n",
      "Processing Batch:  5760 5824\n",
      "Processing Batch:  5824 5888\n",
      "Processing Batch:  5888 5952\n",
      "Processing Batch:  5952 6016\n",
      "Processing Batch:  6016 6080\n",
      "Processing Batch:  6080 6144\n",
      "Processing Batch:  6144 6208\n",
      "Processing Batch:  6208 6272\n",
      "Processing Batch:  6272 6336\n",
      "Processing Batch:  6336 6400\n",
      "Processing Batch:  6400 6464\n",
      "Processing Batch:  6464 6528\n",
      "Processing Batch:  6528 6592\n",
      "Processing Batch:  6592 6656\n",
      "Processing Batch:  6656 6720\n",
      "Processing Batch:  6720 6784\n",
      "Processing Batch:  6784 6848\n",
      "Processing Batch:  6848 6912\n",
      "Processing Batch:  6912 6976\n",
      "Processing Batch:  6976 7040\n",
      "Processing Batch:  7040 7104\n",
      "Processing Batch:  7104 7168\n",
      "Processing Batch:  7168 7232\n",
      "Processing Batch:  7232 7296\n",
      "Processing Batch:  7296 7360\n",
      "Processing Batch:  7360 7424\n",
      "Processing Batch:  7424 7488\n",
      "Processing Batch:  7488 7552\n",
      "Processing Batch:  7552 7616\n",
      "Processing Batch:  7616 7680\n",
      "Processing Batch:  7680 7744\n",
      "Processing Batch:  7744 7808\n",
      "Processing Batch:  7808 7872\n",
      "Processing Batch:  7872 7936\n",
      "Processing Batch:  7936 8000\n",
      "Processing Batch:  8000 8064\n",
      "Processing Batch:  8064 8128\n",
      "Processing Batch:  8128 8192\n",
      "Processing Batch:  8192 8256\n",
      "Processing Batch:  8256 8320\n",
      "Processing Batch:  8320 8384\n",
      "Processing Batch:  8384 8448\n",
      "Processing Batch:  8448 8512\n",
      "Processing Batch:  8512 8576\n",
      "Processing Batch:  8576 8640\n",
      "Processing Batch:  8640 8704\n",
      "Processing Batch:  8704 8768\n",
      "Processing Batch:  8768 8832\n",
      "Processing Batch:  8832 8896\n",
      "Processing Batch:  8896 8960\n",
      "Processing Batch:  8960 9024\n",
      "Processing Batch:  9024 9088\n",
      "Processing Batch:  9088 9152\n",
      "Processing Batch:  9152 9216\n",
      "Processing Batch:  9216 9280\n",
      "Processing Batch:  9280 9344\n",
      "Processing Batch:  9344 9408\n",
      "Processing Batch:  9408 9472\n",
      "Processing Batch:  9472 9536\n",
      "Processing Batch:  9536 9600\n",
      "Processing Batch:  9600 9664\n",
      "Processing Batch:  9664 9728\n",
      "Processing Batch:  9728 9792\n",
      "Processing Batch:  9792 9856\n",
      "Processing Batch:  9856 9920\n",
      "Processing Batch:  9920 9984\n",
      "Processing Batch:  9984 10048\n",
      "Processing Batch:  10048 10112\n",
      "Processing Batch:  10112 10176\n",
      "Processing Batch:  10176 10240\n",
      "Processing Batch:  10240 10304\n",
      "Processing Batch:  10304 10368\n",
      "Processing Batch:  10368 10432\n",
      "Processing Batch:  10432 10496\n",
      "Processing Batch:  10496 10560\n",
      "Processing Batch:  10560 10624\n",
      "Processing Batch:  10624 10688\n",
      "Processing Batch:  10688 10752\n",
      "Processing Batch:  10752 10816\n",
      "Processing Batch:  10816 10880\n",
      "Processing Batch:  10880 10944\n",
      "Processing Batch:  10944 11008\n",
      "Processing Batch:  11008 11072\n",
      "Processing Batch:  11072 11136\n",
      "Processing Batch:  11136 11200\n",
      "Processing Batch:  11200 11264\n",
      "Processing Batch:  11264 11328\n",
      "Processing Batch:  11328 11392\n",
      "Processing Batch:  11392 11456\n",
      "Processing Batch:  11456 11520\n",
      "Processing Batch:  11520 11584\n",
      "Processing Batch:  11584 11648\n",
      "Processing Batch:  11648 11712\n",
      "Processing Batch:  11712 11776\n",
      "Processing Batch:  11776 11840\n",
      "Processing Batch:  11840 11904\n",
      "Processing Batch:  11904 11968\n",
      "Processing Batch:  11968 12032\n",
      "Processing Batch:  12032 12096\n",
      "Processing Batch:  12096 12160\n",
      "Processing Batch:  12160 12224\n",
      "Processing Batch:  12224 12288\n",
      "Processing Batch:  12288 12352\n",
      "Processing Batch:  12352 12416\n",
      "Processing Batch:  12416 12480\n",
      "Processing Batch:  12480 12544\n",
      "Processing Batch:  12544 12608\n",
      "Processing Batch:  12608 12672\n",
      "Processing Batch:  12672 12736\n",
      "Processing Batch:  12736 12800\n",
      "Processing Batch:  12800 12864\n",
      "Processing Batch:  12864 12928\n",
      "Processing Batch:  12928 12992\n",
      "Processing Batch:  12992 13056\n",
      "Processing Batch:  13056 13120\n",
      "Processing Batch:  13120 13184\n",
      "Processing Batch:  13184 13248\n",
      "Processing Batch:  13248 13312\n",
      "Processing Batch:  13312 13376\n",
      "Processing Batch:  13376 13440\n",
      "Processing Batch:  13440 13504\n",
      "Processing Batch:  13504 13568\n",
      "Processing Batch:  13568 13632\n",
      "Processing Batch:  13632 13696\n",
      "Processing Batch:  13696 13760\n",
      "Processing Batch:  13760 13824\n",
      "Processing Batch:  13824 13888\n",
      "Processing Batch:  13888 13952\n",
      "Processing Batch:  13952 14016\n",
      "Processing Batch:  14016 14080\n",
      "Processing Batch:  14080 14144\n",
      "Processing Batch:  14144 14208\n",
      "Processing Batch:  14208 14272\n",
      "Processing Batch:  14272 14336\n",
      "Processing Batch:  14336 14400\n",
      "Processing Batch:  14400 14464\n",
      "Processing Batch:  14464 14528\n",
      "Processing Batch:  14528 14592\n",
      "Processing Batch:  14592 14656\n",
      "Processing Batch:  14656 14720\n",
      "Processing Batch:  14720 14784\n",
      "Processing Batch:  14784 14848\n",
      "Processing Batch:  14848 14912\n",
      "Processing Batch:  14912 14976\n",
      "Processing Batch:  14976 15040\n",
      "Processing Batch:  15040 15104\n",
      "Processing Batch:  15104 15168\n",
      "Processing Batch:  15168 15232\n",
      "Processing Batch:  15232 15296\n",
      "Processing Batch:  15296 15360\n",
      "Processing Batch:  15360 15424\n",
      "Processing Batch:  15424 15488\n",
      "Processing Batch:  15488 15552\n",
      "Processing Batch:  15552 15616\n",
      "Processing Batch:  15616 15680\n",
      "Processing Batch:  15680 15744\n",
      "Processing Batch:  15744 15808\n",
      "Processing Batch:  15808 15872\n",
      "Processing Batch:  15872 15936\n",
      "Processing Batch:  15936 16000\n",
      "Processing Batch:  16000 16064\n",
      "Processing Batch:  16064 16128\n",
      "Processing Batch:  16128 16192\n",
      "Processing Batch:  16192 16256\n",
      "Processing Batch:  16256 16320\n",
      "Processing Batch:  16320 16384\n",
      "Processing Batch:  16384 16448\n",
      "Processing Batch:  16448 16512\n",
      "Processing Batch:  16512 16576\n",
      "Processing Batch:  16576 16640\n",
      "Processing Batch:  16640 16704\n",
      "Processing Batch:  16704 16768\n",
      "Processing Batch:  16768 16832\n",
      "Processing Batch:  16832 16896\n",
      "Processing Batch:  16896 16960\n",
      "Processing Batch:  16960 17024\n",
      "Processing Batch:  17024 17088\n",
      "Processing Batch:  17088 17152\n",
      "Processing Batch:  17152 17216\n",
      "Processing Batch:  17216 17280\n",
      "Processing Batch:  17280 17344\n",
      "Processing Batch:  17344 17408\n",
      "Processing Batch:  17408 17472\n",
      "Processing Batch:  17472 17536\n",
      "Processing Batch:  17536 17600\n",
      "Processing Batch:  17600 17664\n",
      "Processing Batch:  17664 17728\n",
      "Processing Batch:  17728 17792\n",
      "Processing Batch:  17792 17856\n",
      "Processing Batch:  17856 17920\n",
      "Processing Batch:  17920 17984\n",
      "Processing Batch:  17984 18048\n",
      "Processing Batch:  18048 18112\n",
      "Processing Batch:  18112 18176\n",
      "Processing Batch:  18176 18240\n",
      "Processing Batch:  18240 18304\n",
      "Processing Batch:  18304 18368\n",
      "Processing Batch:  18368 18432\n",
      "Processing Batch:  18432 18496\n",
      "Processing Batch:  18496 18560\n",
      "Processing Batch:  18560 18624\n",
      "Processing Batch:  18624 18688\n",
      "Processing Batch:  18688 18752\n",
      "Processing Batch:  18752 18816\n",
      "Processing Batch:  18816 18880\n",
      "Processing Batch:  18880 18944\n",
      "Processing Batch:  18944 19008\n",
      "Processing Batch:  19008 19072\n",
      "Processing Batch:  19072 19136\n",
      "Processing Batch:  19136 19200\n",
      "Processing Batch:  19200 19264\n",
      "Processing Batch:  19264 19328\n",
      "Processing Batch:  19328 19392\n",
      "Processing Batch:  19392 19456\n",
      "Processing Batch:  19456 19520\n",
      "Processing Batch:  19520 19584\n",
      "Processing Batch:  19584 19648\n",
      "Processing Batch:  19648 19712\n",
      "Processing Batch:  19712 19776\n",
      "Processing Batch:  19776 19840\n",
      "Processing Batch:  19840 19904\n",
      "Processing Batch:  19904 19968\n",
      "Processing Batch:  19968 20032\n",
      "Processing Batch:  20032 20096\n",
      "Processing Batch:  20096 20160\n",
      "Processing Batch:  20160 20224\n",
      "Processing Batch:  20224 20288\n",
      "Processing Batch:  20288 20352\n",
      "Processing Batch:  20352 20416\n",
      "Processing Batch:  20416 20480\n",
      "Processing Batch:  20480 20544\n",
      "Processing Batch:  20544 20608\n",
      "Processing Batch:  20608 20672\n",
      "Processing Batch:  20672 20736\n",
      "Processing Batch:  20736 20800\n",
      "Processing Batch:  20800 20864\n",
      "Processing Batch:  20864 20928\n",
      "Processing Batch:  20928 20992\n",
      "Processing Batch:  20992 21056\n",
      "Processing Batch:  21056 21120\n",
      "Processing Batch:  21120 21184\n",
      "Processing Batch:  21184 21248\n",
      "Processing Batch:  21248 21312\n",
      "Processing Batch:  21312 21376\n",
      "Processing Batch:  21376 21440\n",
      "Processing Batch:  21440 21504\n",
      "Processing Batch:  21504 21568\n",
      "Processing Batch:  21568 21632\n",
      "Processing Batch:  21632 21696\n",
      "Processing Batch:  21696 21760\n",
      "Processing Batch:  21760 21824\n",
      "Processing Batch:  21824 21888\n",
      "Processing Batch:  21888 21952\n",
      "Processing Batch:  21952 22016\n",
      "Processing Batch:  22016 22080\n",
      "Processing Batch:  22080 22144\n",
      "Processing Batch:  22144 22208\n",
      "Processing Batch:  22208 22272\n",
      "Processing Batch:  22272 22336\n",
      "Processing Batch:  22336 22400\n",
      "Processing Batch:  22400 22464\n",
      "Processing Batch:  22464 22528\n",
      "Processing Batch:  22528 22592\n",
      "Processing Batch:  22592 22656\n",
      "Processing Batch:  22656 22720\n",
      "Processing Batch:  22720 22784\n",
      "Processing Batch:  22784 22848\n",
      "Processing Batch:  22848 22912\n",
      "Processing Batch:  22912 22976\n",
      "Processing Batch:  22976 23040\n",
      "Processing Batch:  23040 23104\n",
      "Processing Batch:  23104 23168\n",
      "Processing Batch:  23168 23232\n",
      "Processing Batch:  23232 23296\n",
      "Processing Batch:  23296 23360\n",
      "Processing Batch:  23360 23424\n",
      "Processing Batch:  23424 23488\n",
      "Processing Batch:  23488 23552\n",
      "Processing Batch:  23552 23616\n",
      "Processing Batch:  23616 23680\n",
      "Processing Batch:  23680 23744\n",
      "Processing Batch:  23744 23808\n",
      "Processing Batch:  23808 23872\n",
      "Processing Batch:  23872 23936\n",
      "Processing Batch:  23936 24000\n",
      "Processing Batch:  24000 24064\n",
      "Processing Batch:  24064 24128\n",
      "Processing Batch:  24128 24192\n",
      "Processing Batch:  24192 24256\n",
      "Processing Batch:  24256 24320\n",
      "Processing Batch:  24320 24384\n",
      "Processing Batch:  24384 24448\n",
      "Processing Batch:  24448 24512\n",
      "Processing Batch:  24512 24576\n",
      "Processing Batch:  24576 24640\n",
      "Processing Batch:  24640 24704\n",
      "Processing Batch:  24704 24768\n",
      "Processing Batch:  24768 24832\n",
      "Processing Batch:  24832 24896\n",
      "Processing Batch:  24896 24960\n",
      "Processing Batch:  24960 25024\n",
      "Processing Batch:  25024 25088\n",
      "Processing Batch:  25088 25152\n",
      "Processing Batch:  25152 25216\n",
      "Processing Batch:  25216 25280\n",
      "Processing Batch:  25280 25344\n",
      "Processing Batch:  25344 25408\n",
      "Processing Batch:  25408 25472\n",
      "Processing Batch:  25472 25536\n",
      "Processing Batch:  25536 25600\n",
      "Processing Batch:  25600 25664\n",
      "Processing Batch:  25664 25728\n",
      "Processing Batch:  25728 25792\n",
      "Processing Batch:  25792 25856\n",
      "Processing Batch:  25856 25920\n",
      "Processing Batch:  25920 25984\n",
      "Processing Batch:  25984 26048\n",
      "Processing Batch:  26048 26112\n",
      "Processing Batch:  26112 26176\n",
      "Processing Batch:  26176 26240\n",
      "Processing Batch:  26240 26304\n",
      "Processing Batch:  26304 26368\n",
      "Processing Batch:  26368 26432\n",
      "Processing Batch:  26432 26496\n",
      "Processing Batch:  26496 26560\n",
      "Processing Batch:  26560 26624\n",
      "Processing Batch:  26624 26688\n",
      "Processing Batch:  26688 26752\n",
      "Processing Batch:  26752 26816\n",
      "Processing Batch:  26816 26880\n",
      "Processing Batch:  26880 26944\n",
      "Processing Batch:  26944 27008\n",
      "Processing Batch:  27008 27072\n",
      "Processing Batch:  27072 27136\n",
      "Processing Batch:  27136 27200\n",
      "Processing Batch:  27200 27264\n",
      "Processing Batch:  27264 27328\n",
      "Processing Batch:  27328 27392\n",
      "Processing Batch:  27392 27456\n",
      "Processing Batch:  27456 27520\n",
      "Processing Batch:  27520 27584\n",
      "Processing Batch:  27584 27648\n",
      "Processing Batch:  27648 27712\n",
      "Processing Batch:  27712 27776\n",
      "Processing Batch:  27776 27840\n",
      "Processing Batch:  27840 27904\n",
      "Processing Batch:  27904 27968\n",
      "Processing Batch:  27968 28032\n",
      "Processing Batch:  28032 28096\n",
      "Processing Batch:  28096 28160\n",
      "Processing Batch:  28160 28224\n",
      "Processing Batch:  28224 28288\n",
      "Processing Batch:  28288 28352\n",
      "Processing Batch:  28352 28416\n",
      "Processing Batch:  28416 28480\n",
      "Processing Batch:  28480 28544\n",
      "Processing Batch:  28544 28608\n",
      "Processing Batch:  28608 28672\n",
      "Processing Batch:  28672 28736\n",
      "Processing Batch:  28736 28800\n",
      "Processing Batch:  28800 28864\n",
      "Processing Batch:  28864 28928\n",
      "Processing Batch:  28928 28992\n",
      "Processing Batch:  28992 29056\n",
      "Processing Batch:  29056 29120\n",
      "Processing Batch:  29120 29184\n",
      "Processing Batch:  29184 29248\n",
      "Processing Batch:  29248 29312\n",
      "Processing Batch:  29312 29376\n",
      "Processing Batch:  29376 29440\n",
      "Processing Batch:  29440 29504\n",
      "Processing Batch:  29504 29568\n",
      "Processing Batch:  29568 29632\n",
      "Processing Batch:  29632 29696\n",
      "Processing Batch:  29696 29760\n",
      "Processing Batch:  29760 29824\n",
      "Processing Batch:  29824 29888\n",
      "Processing Batch:  29888 29952\n",
      "Processing Batch:  29952 30016\n",
      "Processing Batch:  30016 30080\n",
      "Processing Batch:  30080 30144\n",
      "Processing Batch:  30144 30208\n",
      "Processing Batch:  30208 30272\n",
      "Processing Batch:  30272 30336\n",
      "Processing Batch:  30336 30400\n",
      "Processing Batch:  30400 30464\n",
      "Processing Batch:  30464 30528\n",
      "Processing Batch:  30528 30592\n",
      "Processing Batch:  30592 30656\n",
      "Processing Batch:  30656 30720\n",
      "Processing Batch:  30720 30784\n",
      "Processing Batch:  30784 30848\n",
      "Processing Batch:  30848 30912\n",
      "Processing Batch:  30912 30976\n",
      "Processing Batch:  30976 31040\n",
      "Processing Batch:  31040 31104\n",
      "Processing Batch:  31104 31168\n",
      "Processing Batch:  31168 31232\n",
      "Processing Batch:  31232 31296\n",
      "Processing Batch:  31296 31360\n",
      "Processing Batch:  31360 31424\n",
      "Processing Batch:  31424 31488\n",
      "Processing Batch:  31488 31552\n",
      "Processing Batch:  31552 31616\n",
      "Processing Batch:  31616 31680\n",
      "Processing Batch:  31680 31744\n",
      "Processing Batch:  31744 31808\n",
      "Processing Batch:  31808 31872\n",
      "Processing Batch:  31872 31936\n",
      "Processing Batch:  31936 32000\n",
      "Processing Batch:  32000 32064\n",
      "Processing Batch:  32064 32128\n",
      "Processing Batch:  32128 32192\n",
      "Processing Batch:  32192 32256\n",
      "Processing Batch:  32256 32320\n",
      "Processing Batch:  32320 32384\n",
      "Processing Batch:  32384 32448\n",
      "Processing Batch:  32448 32512\n",
      "Processing Batch:  32512 32576\n",
      "Processing Batch:  32576 32640\n",
      "Processing Batch:  32640 32704\n",
      "Processing Batch:  32704 32768\n",
      "Processing Batch:  32768 32832\n",
      "Processing Batch:  32832 32896\n",
      "Processing Batch:  32896 32960\n",
      "Processing Batch:  32960 33024\n",
      "Processing Batch:  33024 33088\n",
      "Processing Batch:  33088 33152\n",
      "Processing Batch:  33152 33216\n",
      "Processing Batch:  33216 33280\n",
      "Processing Batch:  33280 33344\n",
      "Processing Batch:  33344 33408\n",
      "Processing Batch:  33408 33472\n",
      "Processing Batch:  33472 33536\n",
      "Processing Batch:  33536 33600\n",
      "Processing Batch:  33600 33664\n",
      "Processing Batch:  33664 33728\n",
      "Processing Batch:  33728 33792\n",
      "Processing Batch:  33792 33856\n",
      "Processing Batch:  33856 33920\n",
      "Processing Batch:  33920 33984\n",
      "Processing Batch:  33984 34048\n",
      "Processing Batch:  34048 34112\n",
      "Processing Batch:  34112 34176\n",
      "Processing Batch:  34176 34240\n",
      "Processing Batch:  34240 34304\n",
      "Processing Batch:  34304 34368\n",
      "Processing Batch:  34368 34432\n",
      "Processing Batch:  34432 34496\n",
      "Processing Batch:  34496 34560\n",
      "Processing Batch:  34560 34624\n",
      "Processing Batch:  34624 34688\n",
      "Processing Batch:  34688 34752\n",
      "Processing Batch:  34752 34816\n",
      "Processing Batch:  34816 34880\n",
      "Processing Batch:  34880 34944\n",
      "Processing Batch:  34944 35008\n",
      "Processing Batch:  35008 35072\n",
      "Processing Batch:  35072 35136\n",
      "Processing Batch:  35136 35200\n",
      "Processing Batch:  35200 35264\n",
      "Processing Batch:  35264 35328\n",
      "Processing Batch:  35328 35392\n",
      "Processing Batch:  35392 35456\n",
      "Processing Batch:  35456 35520\n",
      "Processing Batch:  35520 35584\n",
      "Processing Batch:  35584 35648\n",
      "Processing Batch:  35648 35712\n",
      "Processing Batch:  35712 35776\n",
      "Processing Batch:  35776 35840\n",
      "Processing Batch:  35840 35904\n",
      "Processing Batch:  35904 35968\n",
      "Processing Batch:  35968 36032\n",
      "Processing Batch:  36032 36096\n",
      "Processing Batch:  36096 36160\n",
      "Processing Batch:  36160 36224\n",
      "Processing Batch:  36224 36288\n",
      "Processing Batch:  36288 36352\n",
      "Processing Batch:  36352 36416\n",
      "Processing Batch:  36416 36480\n",
      "Processing Batch:  36480 36544\n",
      "Processing Batch:  36544 36608\n",
      "Processing Batch:  36608 36672\n",
      "Processing Batch:  36672 36736\n",
      "Processing Batch:  36736 36800\n",
      "Processing Batch:  36800 36864\n",
      "Processing Batch:  36864 36928\n",
      "Processing Batch:  36928 36992\n",
      "Processing Batch:  36992 37056\n",
      "Processing Batch:  37056 37120\n",
      "Processing Batch:  37120 37184\n",
      "Processing Batch:  37184 37248\n",
      "Processing Batch:  37248 37312\n",
      "Processing Batch:  37312 37376\n",
      "Processing Batch:  37376 37440\n",
      "Processing Batch:  37440 37504\n",
      "Processing Batch:  37504 37568\n",
      "Processing Batch:  37568 37632\n",
      "Processing Batch:  37632 37696\n",
      "Processing Batch:  37696 37760\n",
      "Processing Batch:  37760 37824\n",
      "Processing Batch:  37824 37888\n",
      "Processing Batch:  37888 37952\n",
      "Processing Batch:  37952 38016\n",
      "Processing Batch:  38016 38080\n",
      "Processing Batch:  38080 38144\n",
      "Processing Batch:  38144 38208\n",
      "Processing Batch:  38208 38272\n",
      "Processing Batch:  38272 38336\n",
      "Processing Batch:  38336 38400\n",
      "Processing Batch:  38400 38464\n",
      "Processing Batch:  38464 38528\n",
      "Processing Batch:  38528 38592\n",
      "Processing Batch:  38592 38656\n",
      "Processing Batch:  38656 38720\n",
      "Processing Batch:  38720 38784\n",
      "Processing Batch:  38784 38848\n",
      "Processing Batch:  38848 38912\n",
      "Processing Batch:  38912 38976\n",
      "Processing Batch:  38976 39040\n",
      "Processing Batch:  39040 39104\n",
      "Processing Batch:  39104 39168\n",
      "Processing Batch:  39168 39232\n",
      "Processing Batch:  39232 39296\n",
      "Processing Batch:  39296 39360\n",
      "Processing Batch:  39360 39424\n",
      "Processing Batch:  39424 39488\n",
      "Processing Batch:  39488 39552\n",
      "Processing Batch:  39552 39616\n",
      "Processing Batch:  39616 39680\n",
      "Processing Batch:  39680 39744\n",
      "Processing Batch:  39744 39808\n",
      "Processing Batch:  39808 39872\n",
      "Processing Batch:  39872 39936\n",
      "Processing Batch:  39936 40000\n",
      "Processing Batch:  40000 40064\n",
      "Processing Batch:  40064 40128\n",
      "Processing Batch:  40128 40192\n",
      "Processing Batch:  40192 40256\n",
      "Processing Batch:  40256 40320\n",
      "Processing Batch:  40320 40384\n",
      "Processing Batch:  40384 40448\n",
      "Processing Batch:  40448 40512\n",
      "Processing Batch:  40512 40576\n",
      "Processing Batch:  40576 40640\n",
      "Processing Batch:  40640 40704\n",
      "Processing Batch:  40704 40768\n",
      "Processing Batch:  40768 40832\n",
      "Processing Batch:  40832 40896\n",
      "Processing Batch:  40896 40960\n",
      "Processing Batch:  40960 41024\n",
      "Using GPU: Tesla T4\n",
      "Loading: Rostlab/prot_t5_xl_half_uniref50-enc\n",
      "Model is on device: cuda:0\n",
      "Processing Batch:  0 64\n",
      "Processing Batch:  64 128\n",
      "Processing Batch:  128 192\n",
      "Processing Batch:  192 256\n",
      "Processing Batch:  256 320\n",
      "Processing Batch:  320 384\n",
      "Processing Batch:  384 448\n",
      "Processing Batch:  448 512\n",
      "Processing Batch:  512 576\n",
      "Processing Batch:  576 640\n",
      "Processing Batch:  640 704\n",
      "Processing Batch:  704 768\n",
      "Processing Batch:  768 832\n",
      "Processing Batch:  832 896\n",
      "Processing Batch:  896 960\n",
      "Processing Batch:  960 1024\n",
      "Processing Batch:  1024 1088\n",
      "Processing Batch:  1088 1152\n",
      "Processing Batch:  1152 1216\n",
      "Processing Batch:  1216 1280\n",
      "Processing Batch:  1280 1344\n",
      "Processing Batch:  1344 1408\n",
      "Processing Batch:  1408 1472\n",
      "Processing Batch:  1472 1536\n",
      "Processing Batch:  1536 1600\n",
      "Processing Batch:  1600 1664\n",
      "Processing Batch:  1664 1728\n",
      "Processing Batch:  1728 1792\n",
      "Processing Batch:  1792 1856\n",
      "Processing Batch:  1856 1920\n",
      "Processing Batch:  1920 1984\n",
      "Processing Batch:  1984 2048\n",
      "Processing Batch:  2048 2112\n",
      "Processing Batch:  2112 2176\n",
      "Processing Batch:  2176 2240\n",
      "Processing Batch:  2240 2304\n",
      "Processing Batch:  2304 2368\n",
      "Processing Batch:  2368 2432\n",
      "Processing Batch:  2432 2496\n",
      "Processing Batch:  2496 2560\n",
      "Processing Batch:  2560 2624\n",
      "Processing Batch:  2624 2688\n",
      "Processing Batch:  2688 2752\n",
      "Processing Batch:  2752 2816\n",
      "Processing Batch:  2816 2880\n",
      "Processing Batch:  2880 2944\n",
      "Processing Batch:  2944 3008\n",
      "Processing Batch:  3008 3072\n",
      "Processing Batch:  3072 3136\n",
      "Processing Batch:  3136 3200\n",
      "Processing Batch:  3200 3264\n",
      "Processing Batch:  3264 3328\n",
      "Processing Batch:  3328 3392\n",
      "Processing Batch:  3392 3456\n",
      "Processing Batch:  3456 3520\n",
      "Processing Batch:  3520 3584\n",
      "Processing Batch:  3584 3648\n",
      "Processing Batch:  3648 3712\n",
      "Processing Batch:  3712 3776\n",
      "Processing Batch:  3776 3840\n",
      "Processing Batch:  3840 3904\n",
      "Processing Batch:  3904 3968\n",
      "Processing Batch:  3968 4032\n",
      "Processing Batch:  4032 4096\n",
      "Processing Batch:  4096 4160\n",
      "Processing Batch:  4160 4224\n",
      "Processing Batch:  4224 4288\n",
      "Processing Batch:  4288 4352\n",
      "Processing Batch:  4352 4416\n",
      "Processing Batch:  4416 4480\n",
      "Processing Batch:  4480 4544\n",
      "Processing Batch:  4544 4608\n",
      "Processing Batch:  4608 4672\n",
      "Processing Batch:  4672 4736\n",
      "Processing Batch:  4736 4800\n",
      "Processing Batch:  4800 4864\n",
      "Processing Batch:  4864 4928\n",
      "Processing Batch:  4928 4992\n",
      "Processing Batch:  4992 5056\n",
      "Processing Batch:  5056 5120\n",
      "Processing Batch:  5120 5184\n",
      "Processing Batch:  5184 5248\n",
      "Processing Batch:  5248 5312\n",
      "Processing Batch:  5312 5376\n",
      "Processing Batch:  5376 5440\n",
      "Processing Batch:  5440 5504\n",
      "Processing Batch:  5504 5568\n",
      "Processing Batch:  5568 5632\n",
      "Processing Batch:  5632 5696\n",
      "Processing Batch:  5696 5760\n",
      "Processing Batch:  5760 5824\n",
      "Processing Batch:  5824 5888\n",
      "Processing Batch:  5888 5952\n",
      "Processing Batch:  5952 6016\n",
      "Processing Batch:  6016 6080\n",
      "Processing Batch:  6080 6144\n",
      "Processing Batch:  6144 6208\n",
      "Processing Batch:  6208 6272\n",
      "Processing Batch:  6272 6336\n",
      "Processing Batch:  6336 6400\n",
      "Processing Batch:  6400 6464\n",
      "Processing Batch:  6464 6528\n",
      "Processing Batch:  6528 6592\n",
      "Processing Batch:  6592 6656\n",
      "Processing Batch:  6656 6720\n",
      "Processing Batch:  6720 6784\n",
      "Processing Batch:  6784 6848\n",
      "Processing Batch:  6848 6912\n",
      "Processing Batch:  6912 6976\n",
      "Processing Batch:  6976 7040\n",
      "Processing Batch:  7040 7104\n",
      "Processing Batch:  7104 7168\n",
      "Processing Batch:  7168 7232\n",
      "Processing Batch:  7232 7296\n",
      "Processing Batch:  7296 7360\n",
      "Processing Batch:  7360 7424\n",
      "Processing Batch:  7424 7488\n",
      "Processing Batch:  7488 7552\n",
      "Processing Batch:  7552 7616\n",
      "Processing Batch:  7616 7680\n",
      "Processing Batch:  7680 7744\n",
      "Processing Batch:  7744 7808\n",
      "Processing Batch:  7808 7872\n",
      "Processing Batch:  7872 7936\n",
      "Processing Batch:  7936 8000\n",
      "Processing Batch:  8000 8064\n",
      "Processing Batch:  8064 8128\n",
      "Processing Batch:  8128 8192\n",
      "Processing Batch:  8192 8256\n",
      "Processing Batch:  8256 8320\n",
      "Processing Batch:  8320 8384\n",
      "Processing Batch:  8384 8448\n",
      "Processing Batch:  8448 8512\n",
      "Processing Batch:  8512 8576\n",
      "Processing Batch:  8576 8640\n",
      "Processing Batch:  8640 8704\n",
      "Processing Batch:  8704 8768\n",
      "Processing Batch:  8768 8832\n",
      "Processing Batch:  8832 8896\n",
      "Processing Batch:  8896 8960\n",
      "Processing Batch:  8960 9024\n",
      "Processing Batch:  9024 9088\n",
      "Processing Batch:  9088 9152\n",
      "Processing Batch:  9152 9216\n",
      "Processing Batch:  9216 9280\n",
      "Processing Batch:  9280 9344\n",
      "Processing Batch:  9344 9408\n",
      "Processing Batch:  9408 9472\n",
      "Processing Batch:  9472 9536\n",
      "Processing Batch:  9536 9600\n",
      "Processing Batch:  9600 9664\n",
      "Processing Batch:  9664 9728\n",
      "Processing Batch:  9728 9792\n",
      "Processing Batch:  9792 9856\n",
      "Processing Batch:  9856 9920\n",
      "Processing Batch:  9920 9984\n",
      "Processing Batch:  9984 10048\n",
      "Processing Batch:  10048 10112\n",
      "Processing Batch:  10112 10176\n",
      "Processing Batch:  10176 10240\n",
      "Processing Batch:  10240 10304\n",
      "Processing Batch:  10304 10368\n",
      "Processing Batch:  10368 10432\n",
      "Processing Batch:  10432 10496\n",
      "Processing Batch:  10496 10560\n",
      "Processing Batch:  10560 10624\n",
      "Processing Batch:  10624 10688\n",
      "Processing Batch:  10688 10752\n",
      "Processing Batch:  10752 10816\n",
      "Processing Batch:  10816 10880\n",
      "Processing Batch:  10880 10944\n",
      "Processing Batch:  10944 11008\n",
      "Processing Batch:  11008 11072\n",
      "Processing Batch:  11072 11136\n",
      "Processing Batch:  11136 11200\n",
      "Processing Batch:  11200 11264\n",
      "Processing Batch:  11264 11328\n",
      "Processing Batch:  11328 11392\n",
      "Processing Batch:  11392 11456\n",
      "Processing Batch:  11456 11520\n",
      "Processing Batch:  11520 11584\n",
      "Processing Batch:  11584 11648\n",
      "Processing Batch:  11648 11712\n",
      "Processing Batch:  11712 11776\n",
      "Processing Batch:  11776 11840\n",
      "Processing Batch:  11840 11904\n",
      "Processing Batch:  11904 11968\n",
      "Processing Batch:  11968 12032\n",
      "Processing Batch:  12032 12096\n",
      "Processing Batch:  12096 12160\n",
      "Processing Batch:  12160 12224\n",
      "Processing Batch:  12224 12288\n",
      "Processing Batch:  12288 12352\n",
      "Processing Batch:  12352 12416\n",
      "Processing Batch:  12416 12480\n",
      "Processing Batch:  12480 12544\n",
      "Processing Batch:  12544 12608\n",
      "Processing Batch:  12608 12672\n",
      "Processing Batch:  12672 12736\n",
      "Processing Batch:  12736 12800\n",
      "Processing Batch:  12800 12864\n",
      "Processing Batch:  12864 12928\n",
      "Processing Batch:  12928 12992\n",
      "Processing Batch:  12992 13056\n",
      "Processing Batch:  13056 13120\n",
      "Processing Batch:  13120 13184\n",
      "Processing Batch:  13184 13248\n",
      "Processing Batch:  13248 13312\n",
      "Processing Batch:  13312 13376\n",
      "Processing Batch:  13376 13440\n",
      "Processing Batch:  13440 13504\n",
      "Processing Batch:  13504 13568\n",
      "Processing Batch:  13568 13632\n",
      "Processing Batch:  13632 13696\n",
      "Processing Batch:  13696 13760\n",
      "Processing Batch:  13760 13824\n",
      "Processing Batch:  13824 13888\n",
      "Processing Batch:  13888 13952\n",
      "Processing Batch:  13952 14016\n",
      "Processing Batch:  14016 14080\n",
      "Processing Batch:  14080 14144\n",
      "Processing Batch:  14144 14208\n",
      "Processing Batch:  14208 14272\n",
      "Processing Batch:  14272 14336\n",
      "Processing Batch:  14336 14400\n",
      "Processing Batch:  14400 14464\n",
      "Processing Batch:  14464 14528\n",
      "Processing Batch:  14528 14592\n",
      "Processing Batch:  14592 14656\n",
      "Processing Batch:  14656 14720\n",
      "Processing Batch:  14720 14784\n",
      "Processing Batch:  14784 14848\n",
      "Processing Batch:  14848 14912\n",
      "Processing Batch:  14912 14976\n",
      "Processing Batch:  14976 15040\n",
      "Processing Batch:  15040 15104\n",
      "Processing Batch:  15104 15168\n",
      "Processing Batch:  15168 15232\n",
      "Processing Batch:  15232 15296\n",
      "Processing Batch:  15296 15360\n",
      "Processing Batch:  15360 15424\n",
      "Processing Batch:  15424 15488\n",
      "Processing Batch:  15488 15552\n",
      "Processing Batch:  15552 15616\n",
      "Processing Batch:  15616 15680\n",
      "Processing Batch:  15680 15744\n",
      "Processing Batch:  15744 15808\n",
      "Processing Batch:  15808 15872\n",
      "Processing Batch:  15872 15936\n",
      "Processing Batch:  15936 16000\n",
      "Processing Batch:  16000 16064\n",
      "Processing Batch:  16064 16128\n",
      "Processing Batch:  16128 16192\n",
      "Processing Batch:  16192 16256\n",
      "Processing Batch:  16256 16320\n",
      "Processing Batch:  16320 16384\n",
      "Processing Batch:  16384 16448\n",
      "Processing Batch:  16448 16512\n",
      "Processing Batch:  16512 16576\n",
      "Processing Batch:  16576 16640\n",
      "Processing Batch:  16640 16704\n",
      "Processing Batch:  16704 16768\n",
      "Processing Batch:  16768 16832\n",
      "Processing Batch:  16832 16896\n",
      "Processing Batch:  16896 16960\n",
      "Processing Batch:  16960 17024\n",
      "Processing Batch:  17024 17088\n",
      "Processing Batch:  17088 17152\n",
      "Processing Batch:  17152 17216\n",
      "Processing Batch:  17216 17280\n",
      "Processing Batch:  17280 17344\n",
      "Processing Batch:  17344 17408\n",
      "Processing Batch:  17408 17472\n",
      "Processing Batch:  17472 17536\n",
      "Processing Batch:  17536 17600\n",
      "Processing Batch:  17600 17664\n",
      "Processing Batch:  17664 17728\n",
      "Processing Batch:  17728 17792\n",
      "Processing Batch:  17792 17856\n",
      "Processing Batch:  17856 17920\n",
      "Processing Batch:  17920 17984\n",
      "Processing Batch:  17984 18048\n",
      "Processing Batch:  18048 18112\n",
      "Processing Batch:  18112 18176\n",
      "Processing Batch:  18176 18240\n",
      "Processing Batch:  18240 18304\n",
      "Processing Batch:  18304 18368\n",
      "Processing Batch:  18368 18432\n",
      "Processing Batch:  18432 18496\n",
      "Processing Batch:  18496 18560\n",
      "Processing Batch:  18560 18624\n",
      "Processing Batch:  18624 18688\n",
      "Processing Batch:  18688 18752\n",
      "Processing Batch:  18752 18816\n",
      "Processing Batch:  18816 18880\n",
      "Processing Batch:  18880 18944\n",
      "Processing Batch:  18944 19008\n",
      "Processing Batch:  19008 19072\n",
      "Processing Batch:  19072 19136\n",
      "Processing Batch:  19136 19200\n",
      "Processing Batch:  19200 19264\n",
      "Processing Batch:  19264 19328\n",
      "Processing Batch:  19328 19392\n",
      "Processing Batch:  19392 19456\n",
      "Processing Batch:  19456 19520\n",
      "Processing Batch:  19520 19584\n",
      "Processing Batch:  19584 19648\n",
      "Processing Batch:  19648 19712\n",
      "Processing Batch:  19712 19776\n",
      "Processing Batch:  19776 19840\n",
      "Processing Batch:  19840 19904\n",
      "Processing Batch:  19904 19968\n",
      "Processing Batch:  19968 20032\n",
      "Processing Batch:  20032 20096\n",
      "Processing Batch:  20096 20160\n",
      "Processing Batch:  20160 20224\n",
      "Processing Batch:  20224 20288\n",
      "Processing Batch:  20288 20352\n",
      "Processing Batch:  20352 20416\n",
      "Processing Batch:  20416 20480\n",
      "Processing Batch:  20480 20544\n",
      "Processing Batch:  20544 20608\n",
      "Processing Batch:  20608 20672\n",
      "Processing Batch:  20672 20736\n",
      "Processing Batch:  20736 20800\n",
      "Processing Batch:  20800 20864\n",
      "Processing Batch:  20864 20928\n",
      "Processing Batch:  20928 20992\n",
      "Processing Batch:  20992 21056\n",
      "Processing Batch:  21056 21120\n",
      "Processing Batch:  21120 21184\n",
      "Processing Batch:  21184 21248\n",
      "Processing Batch:  21248 21312\n",
      "Processing Batch:  21312 21376\n",
      "Processing Batch:  21376 21440\n",
      "Processing Batch:  21440 21504\n",
      "Processing Batch:  21504 21568\n",
      "Processing Batch:  21568 21632\n",
      "Processing Batch:  21632 21696\n",
      "Processing Batch:  21696 21760\n",
      "Processing Batch:  21760 21824\n",
      "Processing Batch:  21824 21888\n",
      "Processing Batch:  21888 21952\n",
      "Processing Batch:  21952 22016\n",
      "Processing Batch:  22016 22080\n",
      "Processing Batch:  22080 22144\n",
      "Processing Batch:  22144 22208\n",
      "Processing Batch:  22208 22272\n",
      "Processing Batch:  22272 22336\n",
      "Processing Batch:  22336 22400\n",
      "Processing Batch:  22400 22464\n",
      "Processing Batch:  22464 22528\n",
      "Processing Batch:  22528 22592\n",
      "Processing Batch:  22592 22656\n",
      "Processing Batch:  22656 22720\n",
      "Processing Batch:  22720 22784\n",
      "Processing Batch:  22784 22848\n",
      "Processing Batch:  22848 22912\n",
      "Processing Batch:  22912 22976\n",
      "Processing Batch:  22976 23040\n",
      "Processing Batch:  23040 23104\n",
      "Processing Batch:  23104 23168\n",
      "Processing Batch:  23168 23232\n",
      "Processing Batch:  23232 23296\n",
      "Processing Batch:  23296 23360\n",
      "Processing Batch:  23360 23424\n",
      "Processing Batch:  23424 23488\n",
      "Processing Batch:  23488 23552\n",
      "Processing Batch:  23552 23616\n",
      "Processing Batch:  23616 23680\n",
      "Processing Batch:  23680 23744\n",
      "Processing Batch:  23744 23808\n",
      "Processing Batch:  23808 23872\n",
      "Processing Batch:  23872 23936\n",
      "Processing Batch:  23936 24000\n",
      "Processing Batch:  24000 24064\n",
      "Processing Batch:  24064 24128\n",
      "Processing Batch:  24128 24192\n",
      "Processing Batch:  24192 24256\n",
      "Processing Batch:  24256 24320\n",
      "Processing Batch:  24320 24384\n",
      "Processing Batch:  24384 24448\n",
      "Processing Batch:  24448 24512\n",
      "Processing Batch:  24512 24576\n",
      "Processing Batch:  24576 24640\n",
      "Processing Batch:  24640 24704\n",
      "Processing Batch:  24704 24768\n",
      "Processing Batch:  24768 24832\n",
      "Processing Batch:  24832 24896\n",
      "Processing Batch:  24896 24960\n",
      "Processing Batch:  24960 25024\n",
      "Processing Batch:  25024 25088\n",
      "Processing Batch:  25088 25152\n",
      "Processing Batch:  25152 25216\n",
      "Processing Batch:  25216 25280\n",
      "Processing Batch:  25280 25344\n",
      "Processing Batch:  25344 25408\n",
      "Processing Batch:  25408 25472\n",
      "Processing Batch:  25472 25536\n",
      "Processing Batch:  25536 25600\n",
      "Processing Batch:  25600 25664\n",
      "Processing Batch:  25664 25728\n",
      "Processing Batch:  25728 25792\n",
      "Processing Batch:  25792 25856\n",
      "Processing Batch:  25856 25920\n",
      "Processing Batch:  25920 25984\n",
      "Processing Batch:  25984 26048\n",
      "Processing Batch:  26048 26112\n",
      "Processing Batch:  26112 26176\n",
      "Processing Batch:  26176 26240\n",
      "Processing Batch:  26240 26304\n",
      "Processing Batch:  26304 26368\n",
      "Processing Batch:  26368 26432\n",
      "Processing Batch:  26432 26496\n",
      "Processing Batch:  26496 26560\n",
      "Processing Batch:  26560 26624\n",
      "Processing Batch:  26624 26688\n",
      "Processing Batch:  26688 26752\n",
      "Processing Batch:  26752 26816\n",
      "Processing Batch:  26816 26880\n",
      "Processing Batch:  26880 26944\n",
      "Processing Batch:  26944 27008\n",
      "Processing Batch:  27008 27072\n",
      "Processing Batch:  27072 27136\n",
      "Processing Batch:  27136 27200\n",
      "Processing Batch:  27200 27264\n",
      "Processing Batch:  27264 27328\n",
      "Processing Batch:  27328 27392\n",
      "Processing Batch:  27392 27456\n",
      "Processing Batch:  27456 27520\n",
      "Processing Batch:  27520 27584\n",
      "Processing Batch:  27584 27648\n",
      "Processing Batch:  27648 27712\n",
      "Processing Batch:  27712 27776\n",
      "Processing Batch:  27776 27840\n",
      "Processing Batch:  27840 27904\n",
      "Processing Batch:  27904 27968\n",
      "Processing Batch:  27968 28032\n",
      "Processing Batch:  28032 28096\n",
      "Processing Batch:  28096 28160\n",
      "Processing Batch:  28160 28224\n",
      "Processing Batch:  28224 28288\n",
      "Processing Batch:  28288 28352\n",
      "Processing Batch:  28352 28416\n",
      "Processing Batch:  28416 28480\n",
      "Processing Batch:  28480 28544\n",
      "Processing Batch:  28544 28608\n",
      "Processing Batch:  28608 28672\n",
      "Processing Batch:  28672 28736\n",
      "Processing Batch:  28736 28800\n",
      "Processing Batch:  28800 28864\n",
      "Processing Batch:  28864 28928\n",
      "Processing Batch:  28928 28992\n",
      "Processing Batch:  28992 29056\n",
      "Processing Batch:  29056 29120\n",
      "Processing Batch:  29120 29184\n",
      "Processing Batch:  29184 29248\n",
      "Processing Batch:  29248 29312\n",
      "Processing Batch:  29312 29376\n",
      "Processing Batch:  29376 29440\n",
      "Processing Batch:  29440 29504\n",
      "Processing Batch:  29504 29568\n",
      "Processing Batch:  29568 29632\n",
      "Processing Batch:  29632 29696\n",
      "Processing Batch:  29696 29760\n",
      "Processing Batch:  29760 29824\n",
      "Processing Batch:  29824 29888\n",
      "Processing Batch:  29888 29952\n",
      "Processing Batch:  29952 30016\n",
      "Processing Batch:  30016 30080\n",
      "Processing Batch:  30080 30144\n",
      "Processing Batch:  30144 30208\n",
      "Processing Batch:  30208 30272\n",
      "Processing Batch:  30272 30336\n",
      "Processing Batch:  30336 30400\n",
      "Processing Batch:  30400 30464\n",
      "Processing Batch:  30464 30528\n",
      "Processing Batch:  30528 30592\n",
      "Processing Batch:  30592 30656\n",
      "Processing Batch:  30656 30720\n",
      "Processing Batch:  30720 30784\n",
      "Processing Batch:  30784 30848\n",
      "Processing Batch:  30848 30912\n",
      "Processing Batch:  30912 30976\n",
      "Processing Batch:  30976 31040\n",
      "Processing Batch:  31040 31104\n",
      "Processing Batch:  31104 31168\n",
      "Processing Batch:  31168 31232\n",
      "Processing Batch:  31232 31296\n",
      "Processing Batch:  31296 31360\n",
      "Processing Batch:  31360 31424\n",
      "Processing Batch:  31424 31488\n",
      "Processing Batch:  31488 31552\n",
      "Processing Batch:  31552 31616\n",
      "Processing Batch:  31616 31680\n",
      "Processing Batch:  31680 31744\n",
      "Processing Batch:  31744 31808\n",
      "Processing Batch:  31808 31872\n",
      "Processing Batch:  31872 31936\n",
      "Processing Batch:  31936 32000\n",
      "Processing Batch:  32000 32064\n",
      "Processing Batch:  32064 32128\n",
      "Processing Batch:  32128 32192\n",
      "Processing Batch:  32192 32256\n",
      "Processing Batch:  32256 32320\n",
      "Processing Batch:  32320 32384\n",
      "Processing Batch:  32384 32448\n",
      "Processing Batch:  32448 32512\n",
      "Processing Batch:  32512 32576\n",
      "Processing Batch:  32576 32640\n",
      "Processing Batch:  32640 32704\n",
      "Processing Batch:  32704 32768\n",
      "Processing Batch:  32768 32832\n",
      "Processing Batch:  32832 32896\n",
      "Processing Batch:  32896 32960\n",
      "Processing Batch:  32960 33024\n",
      "Processing Batch:  33024 33088\n",
      "Processing Batch:  33088 33152\n",
      "Processing Batch:  33152 33216\n",
      "Processing Batch:  33216 33280\n",
      "Processing Batch:  33280 33344\n",
      "Processing Batch:  33344 33408\n",
      "Processing Batch:  33408 33472\n",
      "Processing Batch:  33472 33536\n",
      "Processing Batch:  33536 33600\n",
      "Processing Batch:  33600 33664\n",
      "Processing Batch:  33664 33728\n",
      "Processing Batch:  33728 33792\n",
      "Processing Batch:  33792 33856\n",
      "Processing Batch:  33856 33920\n",
      "Processing Batch:  33920 33984\n",
      "Processing Batch:  33984 34048\n",
      "Processing Batch:  34048 34112\n",
      "Processing Batch:  34112 34176\n",
      "Processing Batch:  34176 34240\n",
      "Processing Batch:  34240 34304\n",
      "Processing Batch:  34304 34368\n",
      "Processing Batch:  34368 34432\n",
      "Processing Batch:  34432 34496\n",
      "Processing Batch:  34496 34560\n",
      "Processing Batch:  34560 34624\n",
      "Processing Batch:  34624 34688\n",
      "Processing Batch:  34688 34752\n",
      "Processing Batch:  34752 34816\n",
      "Processing Batch:  34816 34880\n",
      "Processing Batch:  34880 34944\n",
      "Processing Batch:  34944 35008\n",
      "Processing Batch:  35008 35072\n",
      "Processing Batch:  35072 35136\n",
      "Processing Batch:  35136 35200\n",
      "Processing Batch:  35200 35264\n",
      "Processing Batch:  35264 35328\n",
      "Processing Batch:  35328 35392\n",
      "Processing Batch:  35392 35456\n",
      "Processing Batch:  35456 35520\n",
      "Processing Batch:  35520 35584\n",
      "Processing Batch:  35584 35648\n",
      "Processing Batch:  35648 35712\n",
      "Processing Batch:  35712 35776\n",
      "Processing Batch:  35776 35840\n",
      "Processing Batch:  35840 35904\n",
      "Processing Batch:  35904 35968\n",
      "Processing Batch:  35968 36032\n",
      "Processing Batch:  36032 36096\n",
      "Processing Batch:  36096 36160\n",
      "Processing Batch:  36160 36224\n",
      "Processing Batch:  36224 36288\n",
      "Processing Batch:  36288 36352\n",
      "Processing Batch:  36352 36416\n",
      "Processing Batch:  36416 36480\n",
      "Processing Batch:  36480 36544\n",
      "Processing Batch:  36544 36608\n",
      "Processing Batch:  36608 36672\n",
      "Processing Batch:  36672 36736\n",
      "Processing Batch:  36736 36800\n",
      "Processing Batch:  36800 36864\n",
      "Processing Batch:  36864 36928\n",
      "Processing Batch:  36928 36992\n",
      "Processing Batch:  36992 37056\n",
      "Processing Batch:  37056 37120\n",
      "Processing Batch:  37120 37184\n",
      "Processing Batch:  37184 37248\n",
      "Processing Batch:  37248 37312\n",
      "Processing Batch:  37312 37376\n",
      "Processing Batch:  37376 37440\n",
      "Processing Batch:  37440 37504\n",
      "Processing Batch:  37504 37568\n",
      "Processing Batch:  37568 37632\n",
      "Processing Batch:  37632 37696\n",
      "Processing Batch:  37696 37760\n",
      "Processing Batch:  37760 37824\n",
      "Processing Batch:  37824 37888\n",
      "Processing Batch:  37888 37952\n",
      "Processing Batch:  37952 38016\n",
      "Processing Batch:  38016 38080\n",
      "Processing Batch:  38080 38144\n",
      "Processing Batch:  38144 38208\n",
      "Processing Batch:  38208 38272\n",
      "Processing Batch:  38272 38336\n",
      "Processing Batch:  38336 38400\n",
      "Processing Batch:  38400 38464\n",
      "Processing Batch:  38464 38528\n",
      "Processing Batch:  38528 38592\n",
      "Processing Batch:  38592 38656\n",
      "Processing Batch:  38656 38720\n",
      "Processing Batch:  38720 38784\n",
      "Processing Batch:  38784 38848\n",
      "Processing Batch:  38848 38912\n",
      "Processing Batch:  38912 38976\n",
      "Processing Batch:  38976 39040\n",
      "Processing Batch:  39040 39104\n",
      "Processing Batch:  39104 39168\n",
      "Processing Batch:  39168 39232\n",
      "Processing Batch:  39232 39296\n",
      "Processing Batch:  39296 39360\n",
      "Processing Batch:  39360 39424\n",
      "Processing Batch:  39424 39488\n",
      "Processing Batch:  39488 39552\n",
      "Processing Batch:  39552 39616\n",
      "Processing Batch:  39616 39680\n",
      "Processing Batch:  39680 39744\n",
      "Processing Batch:  39744 39808\n",
      "Processing Batch:  39808 39872\n",
      "Processing Batch:  39872 39936\n",
      "Processing Batch:  39936 40000\n",
      "Processing Batch:  40000 40064\n",
      "Processing Batch:  40064 40128\n",
      "Processing Batch:  40128 40192\n",
      "Processing Batch:  40192 40256\n",
      "Processing Batch:  40256 40320\n",
      "Processing Batch:  40320 40384\n",
      "Processing Batch:  40384 40448\n",
      "Processing Batch:  40448 40512\n",
      "Processing Batch:  40512 40576\n",
      "Processing Batch:  40576 40640\n",
      "Processing Batch:  40640 40704\n",
      "Processing Batch:  40704 40768\n",
      "Processing Batch:  40768 40832\n",
      "Processing Batch:  40832 40896\n",
      "Processing Batch:  40896 40960\n",
      "Processing Batch:  40960 41024\n",
      "Processing Batch:  41024 41088\n",
      "Processing Batch:  41088 41152\n",
      "Processing Batch:  41152 41216\n",
      "Processing Batch:  41216 41280\n",
      "Processing Batch:  41280 41344\n",
      "Processing Batch:  41344 41408\n",
      "Processing Batch:  41408 41472\n",
      "Processing Batch:  41472 41536\n",
      "Processing Batch:  41536 41600\n",
      "Processing Batch:  41600 41664\n",
      "Processing Batch:  41664 41728\n",
      "Processing Batch:  41728 41792\n",
      "Processing Batch:  41792 41856\n",
      "Processing Batch:  41856 41920\n",
      "Processing Batch:  41920 41984\n",
      "Processing Batch:  41984 42048\n",
      "Processing Batch:  42048 42112\n",
      "Processing Batch:  42112 42176\n",
      "Processing Batch:  42176 42240\n",
      "Processing Batch:  42240 42304\n",
      "Processing Batch:  42304 42368\n",
      "Processing Batch:  42368 42432\n",
      "Processing Batch:  42432 42496\n",
      "Processing Batch:  42496 42560\n",
      "Processing Batch:  42560 42624\n",
      "Processing Batch:  42624 42688\n",
      "Processing Batch:  42688 42752\n",
      "Processing Batch:  42752 42816\n",
      "Processing Batch:  42816 42880\n",
      "Processing Batch:  42880 42944\n",
      "Processing Batch:  42944 43008\n",
      "Processing Batch:  43008 43072\n",
      "Processing Batch:  43072 43136\n",
      "Processing Batch:  43136 43200\n",
      "Processing Batch:  43200 43264\n",
      "Processing Batch:  43264 43328\n",
      "Processing Batch:  43328 43392\n",
      "Processing Batch:  43392 43456\n",
      "Processing Batch:  43456 43520\n",
      "Processing Batch:  43520 43584\n",
      "Processing Batch:  43584 43648\n",
      "Processing Batch:  43648 43712\n",
      "Processing Batch:  43712 43776\n",
      "Processing Batch:  43776 43840\n",
      "Processing Batch:  43840 43904\n",
      "Processing Batch:  43904 43968\n",
      "Processing Batch:  43968 44032\n",
      "Processing Batch:  44032 44096\n",
      "Processing Batch:  44096 44160\n",
      "Processing Batch:  44160 44224\n",
      "Processing Batch:  44224 44288\n",
      "Processing Batch:  44288 44352\n",
      "Processing Batch:  44352 44416\n",
      "Processing Batch:  44416 44480\n",
      "Processing Batch:  44480 44544\n",
      "Processing Batch:  44544 44608\n",
      "Processing Batch:  44608 44672\n",
      "Using GPU: Tesla T4\n",
      "Loading: Rostlab/prot_t5_xl_half_uniref50-enc\n",
      "Model is on device: cuda:0\n",
      "Processing Batch:  0 64\n",
      "Processing Batch:  64 128\n",
      "Processing Batch:  128 192\n",
      "Processing Batch:  192 256\n",
      "Processing Batch:  256 320\n",
      "Processing Batch:  320 384\n",
      "Processing Batch:  384 448\n",
      "Processing Batch:  448 512\n",
      "Processing Batch:  512 576\n",
      "Processing Batch:  576 640\n",
      "Processing Batch:  640 704\n",
      "Processing Batch:  704 768\n",
      "Processing Batch:  768 832\n",
      "Processing Batch:  832 896\n",
      "Processing Batch:  896 960\n",
      "Processing Batch:  960 1024\n",
      "Processing Batch:  1024 1088\n",
      "Processing Batch:  1088 1152\n",
      "Processing Batch:  1152 1216\n",
      "Processing Batch:  1216 1280\n",
      "Processing Batch:  1280 1344\n",
      "Processing Batch:  1344 1408\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_53158/1333535040.py:25: DtypeWarning: Columns (0,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df_beta_train = pd.read_csv(path_beta_train, sep=\"\\t\", index_col=False)\n"
     ]
    }
   ],
   "source": [
    "path_paired_test = f\"data_10x/splitted_datasets/{precision}/paired/test.tsv\"\n",
    "path_paired_validation = f\"data_10x/splitted_datasets/{precision}/paired/validation.tsv\"\n",
    "path_paired_train = f\"data_10x/splitted_datasets/{precision}/paired/train.tsv\"\n",
    "path_beta_test = f\"data_10x/splitted_datasets/{precision}/beta/test.tsv\"\n",
    "path_beta_validation = f\"data_10x/splitted_datasets/{precision}/beta/validation.tsv\"\n",
    "path_beta_train = f\"data_10x/splitted_datasets/{precision}/beta/train.tsv\"\n",
    "\n",
    "path_paired = f\"{pipeline_data}/embeddings/temp/{precision}/paired_concatenated.tsv\"\n",
    "create_folders_if_not_exists([os.path.dirname(path_paired)])\n",
    "df_paired_test = pd.read_csv(path_paired_test, sep=\"\\t\", index_col=False)\n",
    "df_paired_validation = pd.read_csv(path_paired_validation, sep=\"\\t\", index_col=False)\n",
    "df_paired_train = pd.read_csv(path_paired_train, sep=\"\\t\", index_col=False)\n",
    "df_paired = pd.concat([df_paired_test, df_paired_validation, df_paired_train])\n",
    "df_paired.to_csv(path_paired, sep=\"\\t\", index=False)\n",
    "\n",
    "# paired\n",
    "%run ./data_scripts/generateEmbeddings.py paired {path_paired} {pipeline_data}/embeddings/paired/{precision}/TRA_paired_embeddings.npz TRA_CDR3\n",
    "%run ./data_scripts/generateEmbeddings.py paired {path_paired} {pipeline_data}/embeddings/paired/{precision}/TRB_paired_embeddings.npz TRB_CDR3\n",
    "%run ./data_scripts/generateEmbeddings.py paired {path_paired} {pipeline_data}/embeddings/paired/{precision}/Epitope_paired_embeddings.npz Epitope\n",
    "\n",
    "path_beta = f\"{pipeline_data}/embeddings/temp/{precision}/beta_concatenated.tsv\"\n",
    "create_folders_if_not_exists([os.path.dirname(path_beta)])\n",
    "df_beta_test = pd.read_csv(path_beta_test, sep=\"\\t\", index_col=False)\n",
    "df_beta_validation = pd.read_csv(path_beta_validation, sep=\"\\t\", index_col=False)\n",
    "df_beta_train = pd.read_csv(path_beta_train, sep=\"\\t\", index_col=False)\n",
    "df_beta = pd.concat([df_beta_test, df_beta_validation, df_beta_train])\n",
    "df_beta.to_csv(path_beta, sep=\"\\t\", index=False)\n",
    "\n",
    "# beta\n",
    "#%run ./data_scripts/generateEmbeddings.py beta {path_beta} {pipeline_data}/embeddings/beta/{precision}/TRB_beta_embeddings.npz TRB_CDR3\n",
    "#%run ./data_scripts/generateEmbeddings.py beta {path_beta} {pipeline_data}/embeddings/beta/{precision}/Epitope_beta_embeddings.npz Epitope"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "from transformers import T5Tokenizer, T5EncoderModel\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create Physicochemical Properties"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/home/ubuntu/PA-Cancer-Immunotherapy-Transformer/BA_ZHAW/./data_scripts/generatePhysicoParallel.py\", line 1, in <module>\n",
      "    import peptides\n",
      "ModuleNotFoundError: No module named 'peptides'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/ubuntu/PA-Cancer-Immunotherapy-Transformer/BA_ZHAW/./data_scripts/generatePhysicoParallel.py\", line 1, in <module>\n",
      "    import peptides\n",
      "ModuleNotFoundError: No module named 'peptides'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/ubuntu/PA-Cancer-Immunotherapy-Transformer/BA_ZHAW/./data_scripts/generatePhysicoParallel.py\", line 1, in <module>\n",
      "    import peptides\n",
      "ModuleNotFoundError: No module named 'peptides'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/ubuntu/PA-Cancer-Immunotherapy-Transformer/BA_ZHAW/./data_scripts/generatePhysicoParallel.py\", line 1, in <module>\n",
      "    import peptides\n",
      "ModuleNotFoundError: No module named 'peptides'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/ubuntu/PA-Cancer-Immunotherapy-Transformer/BA_ZHAW/./data_scripts/generatePhysicoParallel.py\", line 1, in <module>\n",
      "    import peptides\n",
      "ModuleNotFoundError: No module named 'peptides'\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/ubuntu/PA-Cancer-Immunotherapy-Transformer/BA_ZHAW/./data_scripts/generatePhysicoParallel.py\", line 1, in <module>\n",
      "    import peptides\n",
      "ModuleNotFoundError: No module named 'peptides'\n"
     ]
    }
   ],
   "source": [
    "!python ./data_scripts/generatePhysicoParallel.py paired {pipeline_data_splitted}/{precision}/paired test ./data_10x/physicoProperties {precision}\n",
    "!python ./data_scripts/generatePhysicoParallel.py paired {pipeline_data_splitted}/{precision}/paired validation ./data_10x/physicoProperties {precision}\n",
    "!python ./data_scripts/generatePhysicoParallel.py paired {pipeline_data_splitted}/{precision}/paired train ./data_10x/physicoProperties {precision}\n",
    "\n",
    "!python ./data_scripts/generatePhysicoParallel.py beta {pipeline_data_splitted}/{precision}/beta test ./data_10x/physicoProperties {precision}\n",
    "!python ./data_scripts/generatePhysicoParallel.py beta {pipeline_data_splitted}/{precision}/beta validation ./data_10x/physicoProperties {precision}\n",
    "!python ./data_scripts/generatePhysicoParallel.py beta {pipeline_data_splitted}/{precision}/beta train ./data_10x/physicoProperties {precision}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Scale Physicochemical Properties"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'peptides'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "File \u001b[0;32m/tmp/ipykernel_53158/2556338357.py:5\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[38;5;66;03m# from sklearn.preprocessing import StandardScaler\u001b[39;00m\n\u001b[1;32m      4\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01msklearn\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mpreprocessing\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m MinMaxScaler\n\u001b[0;32m----> 5\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mpeptides\u001b[39;00m\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'peptides'"
     ]
    },
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'peptides'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[53], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m base_path \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m./data_10x/physicoProperties\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m      2\u001b[0m chain \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpaired\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m----> 3\u001b[0m get_ipython()\u001b[38;5;241m.\u001b[39mrun_line_magic(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mrun\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m./data_scripts/scale_physicos.ipynb\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m      5\u001b[0m chain \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mbeta\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m      6\u001b[0m get_ipython()\u001b[38;5;241m.\u001b[39mrun_line_magic(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mrun\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m./data_scripts/scale_physicos.ipynb\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.12/site-packages/IPython/core/interactiveshell.py:2480\u001b[0m, in \u001b[0;36mInteractiveShell.run_line_magic\u001b[0;34m(self, magic_name, line, _stack_depth)\u001b[0m\n\u001b[1;32m   2478\u001b[0m     kwargs[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mlocal_ns\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mget_local_scope(stack_depth)\n\u001b[1;32m   2479\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mbuiltin_trap:\n\u001b[0;32m-> 2480\u001b[0m     result \u001b[38;5;241m=\u001b[39m fn(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[1;32m   2482\u001b[0m \u001b[38;5;66;03m# The code below prevents the output from being displayed\u001b[39;00m\n\u001b[1;32m   2483\u001b[0m \u001b[38;5;66;03m# when using magics with decorator @output_can_be_silenced\u001b[39;00m\n\u001b[1;32m   2484\u001b[0m \u001b[38;5;66;03m# when the last Python token in the expression is a ';'.\u001b[39;00m\n\u001b[1;32m   2485\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mgetattr\u001b[39m(fn, magic\u001b[38;5;241m.\u001b[39mMAGIC_OUTPUT_CAN_BE_SILENCED, \u001b[38;5;28;01mFalse\u001b[39;00m):\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.12/site-packages/IPython/core/magics/execution.py:737\u001b[0m, in \u001b[0;36mExecutionMagics.run\u001b[0;34m(self, parameter_s, runner, file_finder)\u001b[0m\n\u001b[1;32m    735\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m preserve_keys(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mshell\u001b[38;5;241m.\u001b[39muser_ns, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m__file__\u001b[39m\u001b[38;5;124m'\u001b[39m):\n\u001b[1;32m    736\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mshell\u001b[38;5;241m.\u001b[39muser_ns[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m__file__\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m filename\n\u001b[0;32m--> 737\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mshell\u001b[38;5;241m.\u001b[39msafe_execfile_ipy(filename, raise_exceptions\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[1;32m    738\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m\n\u001b[1;32m    740\u001b[0m \u001b[38;5;66;03m# Control the response to exit() calls made by the script being run\u001b[39;00m\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.12/site-packages/IPython/core/interactiveshell.py:3005\u001b[0m, in \u001b[0;36mInteractiveShell.safe_execfile_ipy\u001b[0;34m(self, fname, shell_futures, raise_exceptions)\u001b[0m\n\u001b[1;32m   3003\u001b[0m result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mrun_cell(cell, silent\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m, shell_futures\u001b[38;5;241m=\u001b[39mshell_futures)\n\u001b[1;32m   3004\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m raise_exceptions:\n\u001b[0;32m-> 3005\u001b[0m     result\u001b[38;5;241m.\u001b[39mraise_error()\n\u001b[1;32m   3006\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m result\u001b[38;5;241m.\u001b[39msuccess:\n\u001b[1;32m   3007\u001b[0m     \u001b[38;5;28;01mbreak\u001b[39;00m\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.12/site-packages/IPython/core/interactiveshell.py:308\u001b[0m, in \u001b[0;36mExecutionResult.raise_error\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    306\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39merror_before_exec\n\u001b[1;32m    307\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39merror_in_exec \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m--> 308\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39merror_in_exec\n",
      "    \u001b[0;31m[... skipping hidden 1 frame]\u001b[0m\n",
      "File \u001b[0;32m/tmp/ipykernel_53158/2556338357.py:5\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[38;5;66;03m# from sklearn.preprocessing import StandardScaler\u001b[39;00m\n\u001b[1;32m      4\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01msklearn\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mpreprocessing\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m MinMaxScaler\n\u001b[0;32m----> 5\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mpeptides\u001b[39;00m\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'peptides'"
     ]
    }
   ],
   "source": [
    "base_path = \"./data_10x/physicoProperties\"\n",
    "chain = \"paired\"\n",
    "%run ./data_scripts/scale_physicos.ipynb\n",
    "\n",
    "chain = \"beta\"\n",
    "%run ./data_scripts/scale_physicos.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_103149/1419232561.py:34: DtypeWarning: Columns (0,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  train_df = pd.read_csv(paths[\"train\"], sep='\\t')\n",
      "/tmp/ipykernel_103149/1419232561.py:34: DtypeWarning: Columns (0,4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  train_df = pd.read_csv(paths[\"train\"], sep='\\t')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--- Paired Gene ---\n",
      "Anzahl der Zeilen im Trainingsdatensatz: 61498 (Binding=1: 33711, Binding=0: 27787, TPP1: 0, TPP2: 0, TPP3: 0)\n",
      "Anzahl der Zeilen im Testdatensatz: 36806 (Binding=1: 7226, Binding=0: 29580, TPP1: 22545, TPP2: 13972, TPP3: 289)\n",
      "Anzahl der Zeilen im Validierungsdatensatz: 36934 (Binding=1: 7224, Binding=0: 29710, TPP1: 0, TPP2: 0, TPP3: 0)\n",
      "Gesamtanzahl der Zeilen (Train + Test + Validation): 135238\n",
      "\n",
      "--- Paired Allele ---\n",
      "Anzahl der Zeilen im Trainingsdatensatz: 66590 (Binding=1: 36515, Binding=0: 30075, TPP1: 0, TPP2: 0, TPP3: 0)\n",
      "Anzahl der Zeilen im Testdatensatz: 39891 (Binding=1: 7826, Binding=0: 32065, TPP1: 25231, TPP2: 14579, TPP3: 81)\n",
      "Anzahl der Zeilen im Validierungsdatensatz: 40021 (Binding=1: 7826, Binding=0: 32195, TPP1: 0, TPP2: 0, TPP3: 0)\n",
      "Gesamtanzahl der Zeilen (Train + Test + Validation): 146502\n",
      "\n",
      "--- Beta Gene ---\n",
      "Anzahl der Zeilen im Trainingsdatensatz: 251750 (Binding=1: 125875, Binding=0: 125875, TPP1: 0, TPP2: 0, TPP3: 0)\n",
      "Anzahl der Zeilen im Testdatensatz: 161844 (Binding=1: 26974, Binding=0: 134870, TPP1: 140580, TPP2: 20961, TPP3: 298)\n",
      "Anzahl der Zeilen im Validierungsdatensatz: 161838 (Binding=1: 26973, Binding=0: 134865, TPP1: 0, TPP2: 0, TPP3: 0)\n",
      "Gesamtanzahl der Zeilen (Train + Test + Validation): 575432\n",
      "\n",
      "--- Beta Allele ---\n",
      "Anzahl der Zeilen im Trainingsdatensatz: 279290 (Binding=1: 139645, Binding=0: 139645, TPP1: 0, TPP2: 0, TPP3: 0)\n",
      "Anzahl der Zeilen im Testdatensatz: 179550 (Binding=1: 29925, Binding=0: 149625, TPP1: 157873, TPP2: 21581, TPP3: 94)\n",
      "Anzahl der Zeilen im Validierungsdatensatz: 179532 (Binding=1: 29922, Binding=0: 149610, TPP1: 0, TPP2: 0, TPP3: 0)\n",
      "Gesamtanzahl der Zeilen (Train + Test + Validation): 638372\n",
      "\n",
      "         Dataset   Train  Train_Binding_1  Train_Binding_0  Train_TPP1  \\\n",
      "0    Paired Gene   61498            33711            27787           0   \n",
      "1  Paired Allele   66590            36515            30075           0   \n",
      "2      Beta Gene  251750           125875           125875           0   \n",
      "3    Beta Allele  279290           139645           139645           0   \n",
      "\n",
      "   Train_TPP2  Train_TPP3  Train_TPP4    Test  Test_Binding_1  ...  Test_TPP3  \\\n",
      "0           0           0           0   36806            7226  ...        289   \n",
      "1           0           0           0   39891            7826  ...         81   \n",
      "2           0           0           0  161844           26974  ...        298   \n",
      "3           0           0           0  179550           29925  ...         94   \n",
      "\n",
      "   Test_TPP4  Validation  Validation_Binding_1  Validation_Binding_0  \\\n",
      "0          0       36934                  7224                 29710   \n",
      "1          0       40021                  7826                 32195   \n",
      "2          5      161838                 26973                134865   \n",
      "3          2      179532                 29922                149610   \n",
      "\n",
      "   Validation_TPP1  Validation_TPP2  Validation_TPP3  Validation_TPP4   Total  \n",
      "0                0                0                0                0  135238  \n",
      "1                0                0                0                0  146502  \n",
      "2                0                0                0                0  575432  \n",
      "3                0                0                0                0  638372  \n",
      "\n",
      "[4 rows x 23 columns]\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Beispielpfade für Train-, Test-, und Validierungsdatensätze für alle vier Kategorien\n",
    "base_path = 'data_10x/splitted_datasets'\n",
    "\n",
    "# Definierte Pfade für alle vier Kategorien\n",
    "datasets = {\n",
    "    \"paired_gene\": {\n",
    "        \"train\": f\"{base_path}/gene/paired/train.tsv\",\n",
    "        \"test\": f\"{base_path}/gene/paired/test.tsv\",\n",
    "        \"validation\": f\"{base_path}/gene/paired/validation.tsv\"\n",
    "    },\n",
    "    \"paired_allele\": {\n",
    "        \"train\": f\"{base_path}/allele/paired/train.tsv\",\n",
    "        \"test\": f\"{base_path}/allele/paired/test.tsv\",\n",
    "        \"validation\": f\"{base_path}/allele/paired/validation.tsv\"\n",
    "    },\n",
    "    \"beta_gene\": {\n",
    "        \"train\": f\"{base_path}/gene/beta/train.tsv\",\n",
    "        \"test\": f\"{base_path}/gene/beta/test.tsv\",\n",
    "        \"validation\": f\"{base_path}/gene/beta/validation.tsv\"\n",
    "    },\n",
    "    \"beta_allele\": {\n",
    "        \"train\": f\"{base_path}/allele/beta/train.tsv\",\n",
    "        \"test\": f\"{base_path}/allele/beta/test.tsv\",\n",
    "        \"validation\": f\"{base_path}/allele/beta/validation.tsv\"\n",
    "    }\n",
    "}\n",
    "\n",
    "# Berechnung der Anzahl der Zeilen für jedes Set\n",
    "results = {}\n",
    "for dataset_name, paths in datasets.items():\n",
    "    # Daten laden\n",
    "    train_df = pd.read_csv(paths[\"train\"], sep='\\t')\n",
    "    test_df = pd.read_csv(paths[\"test\"], sep='\\t')\n",
    "    validation_df = pd.read_csv(paths[\"validation\"], sep='\\t')\n",
    "    \n",
    "    # Anzahl der Zeilen berechnen\n",
    "    train_length = len(train_df)\n",
    "    test_length = len(test_df)\n",
    "    validation_length = len(validation_df)\n",
    "    total_length = train_length + test_length + validation_length\n",
    "    \n",
    "    # Zähle die Anzahl der Bindings 1 und 0 in jedem Datensatz\n",
    "    train_binding_counts = train_df['Binding'].value_counts()\n",
    "    test_binding_counts = test_df['Binding'].value_counts()\n",
    "    validation_binding_counts = validation_df['Binding'].value_counts()\n",
    "    \n",
    "    # Zähle die Anzahl der TPP1, TPP2, TPP3 Einträge in jedem Datensatz\n",
    "    train_task_counts = train_df['task'].value_counts()\n",
    "    test_task_counts = test_df['task'].value_counts()\n",
    "    validation_task_counts = validation_df['task'].value_counts()\n",
    "\n",
    "    # Ergebnisse speichern\n",
    "    results[dataset_name] = {\n",
    "        \"Train\": train_length,\n",
    "        \"Train_Binding_1\": train_binding_counts.get(1, 0),\n",
    "        \"Train_Binding_0\": train_binding_counts.get(0, 0),\n",
    "        \"Train_TPP1\": train_task_counts.get(\"TPP1\", 0),\n",
    "        \"Train_TPP2\": train_task_counts.get(\"TPP2\", 0),\n",
    "        \"Train_TPP3\": train_task_counts.get(\"TPP3\", 0),\n",
    "        \"Train_TPP4\": train_task_counts.get(\"TPP4\", 0),\n",
    "        \"Test\": test_length,\n",
    "        \"Test_Binding_1\": test_binding_counts.get(1, 0),\n",
    "        \"Test_Binding_0\": test_binding_counts.get(0, 0),\n",
    "        \"Test_TPP1\": test_task_counts.get(\"TPP1\", 0),\n",
    "        \"Test_TPP2\": test_task_counts.get(\"TPP2\", 0),\n",
    "        \"Test_TPP3\": test_task_counts.get(\"TPP3\", 0),\n",
    "        \"Test_TPP4\": test_task_counts.get(\"TPP4\", 0),\n",
    "        \"Validation\": validation_length,\n",
    "        \"Validation_Binding_1\": validation_binding_counts.get(1, 0),\n",
    "        \"Validation_Binding_0\": validation_binding_counts.get(0, 0),\n",
    "        \"Validation_TPP1\": validation_task_counts.get(\"TPP1\", 0),\n",
    "        \"Validation_TPP2\": validation_task_counts.get(\"TPP2\", 0),\n",
    "        \"Validation_TPP3\": validation_task_counts.get(\"TPP3\", 0),\n",
    "        \"Validation_TPP4\": validation_task_counts.get(\"TPP4\", 0),\n",
    "        \"Total\": total_length\n",
    "    }\n",
    "\n",
    "# Ergebnisse anzeigen\n",
    "for dataset, lengths in results.items():\n",
    "    print(f'--- {dataset.replace(\"_\", \" \").title()} ---')\n",
    "    print(f'Anzahl der Zeilen im Trainingsdatensatz: {lengths[\"Train\"]} (Binding=1: {lengths[\"Train_Binding_1\"]}, Binding=0: {lengths[\"Train_Binding_0\"]}, TPP1: {lengths[\"Train_TPP1\"]}, TPP2: {lengths[\"Train_TPP2\"]}, TPP3: {lengths[\"Train_TPP3\"]})')\n",
    "    print(f'Anzahl der Zeilen im Testdatensatz: {lengths[\"Test\"]} (Binding=1: {lengths[\"Test_Binding_1\"]}, Binding=0: {lengths[\"Test_Binding_0\"]}, TPP1: {lengths[\"Test_TPP1\"]}, TPP2: {lengths[\"Test_TPP2\"]}, TPP3: {lengths[\"Test_TPP3\"]})')\n",
    "    print(f'Anzahl der Zeilen im Validierungsdatensatz: {lengths[\"Validation\"]} (Binding=1: {lengths[\"Validation_Binding_1\"]}, Binding=0: {lengths[\"Validation_Binding_0\"]}, TPP1: {lengths[\"Validation_TPP1\"]}, TPP2: {lengths[\"Validation_TPP2\"]}, TPP3: {lengths[\"Validation_TPP3\"]})')\n",
    "    print(f'Gesamtanzahl der Zeilen (Train + Test + Validation): {lengths[\"Total\"]}\\n')\n",
    "\n",
    "# Optional: Ergebnisse in einer Übersichtstabelle darstellen\n",
    "summary_data = []\n",
    "for dataset, lengths in results.items():\n",
    "    summary_data.append({\n",
    "        \"Dataset\": dataset.replace(\"_\", \" \").title(),\n",
    "        \"Train\": lengths[\"Train\"],\n",
    "        \"Train_Binding_1\": lengths[\"Train_Binding_1\"],\n",
    "        \"Train_Binding_0\": lengths[\"Train_Binding_0\"],\n",
    "        \"Train_TPP1\": lengths[\"Train_TPP1\"],\n",
    "        \"Train_TPP2\": lengths[\"Train_TPP2\"],\n",
    "        \"Train_TPP3\": lengths[\"Train_TPP3\"],\n",
    "        \"Train_TPP4\": lengths[\"Train_TPP4\"],\n",
    "        \"Test\": lengths[\"Test\"],\n",
    "        \"Test_Binding_1\": lengths[\"Test_Binding_1\"],\n",
    "        \"Test_Binding_0\": lengths[\"Test_Binding_0\"],\n",
    "        \"Test_TPP1\": lengths[\"Test_TPP1\"],\n",
    "        \"Test_TPP2\": lengths[\"Test_TPP2\"],\n",
    "        \"Test_TPP3\": lengths[\"Test_TPP3\"],\n",
    "        \"Test_TPP4\": lengths[\"Test_TPP4\"],\n",
    "        \"Validation\": lengths[\"Validation\"],\n",
    "        \"Validation_Binding_1\": lengths[\"Validation_Binding_1\"],\n",
    "        \"Validation_Binding_0\": lengths[\"Validation_Binding_0\"],\n",
    "        \"Validation_TPP1\": lengths[\"Validation_TPP1\"],\n",
    "        \"Validation_TPP2\": lengths[\"Validation_TPP2\"],\n",
    "        \"Validation_TPP3\": lengths[\"Validation_TPP3\"],\n",
    "        \"Validation_TPP4\": lengths[\"Validation_TPP4\"],\n",
    "        \"Total\": lengths[\"Total\"]\n",
    "    })\n",
    "\n",
    "summary_df = pd.DataFrame(summary_data)\n",
    "print(summary_df)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
